// @fetchai/blst-ts/prebuild/emscripten/blst.js


// The Module object: Our interface to the outside world. We import
// and export values on it. There are various ways Module can be used:
// 1. Not defined. We create it here
// 2. A function parameter, function(Module) { ..generated code.. }
// 3. pre-run appended it, var Module = {}; ..generated code..
// 4. External script tag defines var Module.
// We need to check if Module already exists (e.g. case 3 above).
// Substitution will be replaced with actual code on later stage of the build,
// this way Closure Compiler will not mangle it (e.g. case 4. above).
// Note that if you want to run closure, and also to use Module
// after the generated code, you will need to define   var Module = {};
// before the code. Then that object will be used in the code, and you
// can continue to use Module afterwards as well.
var Module = typeof Module != 'undefined' ? Module : {};

// See https://caniuse.com/mdn-javascript_builtins_object_assign

// --pre-jses are emitted after the Module integration code, so that they can
// refer to Module (if they choose; they can also define Module)
Module['wasmBinary'] = require("./blst.wasm.js").default;


// Sometimes an existing Module object exists with properties
// meant to overwrite the default module functionality. Here
// we collect those properties and reapply _after_ we configure
// the current environment's defaults to avoid having to be so
// defensive during initialization.
var moduleOverrides = Object.assign({}, Module);

var arguments_ = [];
var thisProgram = './this.program';
var quit_ = (status, toThrow) => {
  throw toThrow;
};

// Determine the runtime environment we are in. You can customize this by
// setting the ENVIRONMENT setting at compile time (see settings.js).

// Attempt to auto-detect the environment
var ENVIRONMENT_IS_WEB = typeof window == 'object';
var ENVIRONMENT_IS_WORKER = typeof importScripts == 'function';
// N.b. Electron.js environment is simultaneously a NODE-environment, but
// also a web environment.
var ENVIRONMENT_IS_NODE = typeof process == 'object' && typeof process.versions == 'object' && typeof process.versions.node == 'string';
var ENVIRONMENT_IS_SHELL = !ENVIRONMENT_IS_WEB && !ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER;

if (Module['ENVIRONMENT']) {
  throw new Error('Module.ENVIRONMENT has been deprecated. To force the environment, use the ENVIRONMENT compile-time option (for example, -sENVIRONMENT=web or -sENVIRONMENT=node)');
}

// `/` should be present at the end if `scriptDirectory` is not empty
var scriptDirectory = '';
function locateFile(path) {
  if (Module['locateFile']) {
    return Module['locateFile'](path, scriptDirectory);
  }
  return scriptDirectory + path;
}

// Hooks that are implemented differently in different runtime environments.
var read_,
    readAsync,
    readBinary,
    setWindowTitle;

// Normally we don't log exceptions but instead let them bubble out the top
// level where the embedding environment (e.g. the browser) can handle
// them.
// However under v8 and node we sometimes exit the process direcly in which case
// its up to use us to log the exception before exiting.
// If we fix https://github.com/emscripten-core/emscripten/issues/15080
// this may no longer be needed under node.
function logExceptionOnExit(e) {
  if (e instanceof ExitStatus) return;
  let toLog = e;
  if (e && typeof e == 'object' && e.stack) {
    toLog = [e, e.stack];
  }
  err('exiting due to exception: ' + toLog);
}

var fs;
var nodePath;
var requireNodeFS;

if (ENVIRONMENT_IS_NODE) {
  if (!(typeof process == 'object' && typeof require == 'function')) throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)');
  if (ENVIRONMENT_IS_WORKER) {
    scriptDirectory = require('path').dirname(scriptDirectory) + '/';
  } else {
    scriptDirectory = __dirname + '/';
  }

// include: node_shell_read.js


requireNodeFS = () => {
  // Use nodePath as the indicator for these not being initialized,
  // since in some environments a global fs may have already been
  // created.
  if (!nodePath) {
    fs = require('fs');
    nodePath = require('path');
  }
};

read_ = function shell_read(filename, binary) {
  requireNodeFS();
  filename = nodePath['normalize'](filename);
  return fs.readFileSync(filename, binary ? undefined : 'utf8');
};

readBinary = (filename) => {
  var ret = read_(filename, true);
  if (!ret.buffer) {
    ret = new Uint8Array(ret);
  }
  assert(ret.buffer);
  return ret;
};

readAsync = (filename, onload, onerror) => {
  requireNodeFS();
  filename = nodePath['normalize'](filename);
  fs.readFile(filename, function(err, data) {
    if (err) onerror(err);
    else onload(data.buffer);
  });
};

// end include: node_shell_read.js
  if (process['argv'].length > 1) {
    thisProgram = process['argv'][1].replace(/\\/g, '/');
  }

  arguments_ = process['argv'].slice(2);

  if (typeof module != 'undefined') {
    module['exports'] = Module;
  }

  process['on']('uncaughtException', function(ex) {
    // suppress ExitStatus exceptions from showing an error
    if (!(ex instanceof ExitStatus)) {
      throw ex;
    }
  });

  // Without this older versions of node (< v15) will log unhandled rejections
  // but return 0, which is not normally the desired behaviour.  This is
  // not be needed with node v15 and about because it is now the default
  // behaviour:
  // See https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode
  process['on']('unhandledRejection', function(reason) { throw reason; });

  quit_ = (status, toThrow) => {
    if (keepRuntimeAlive()) {
      process['exitCode'] = status;
      throw toThrow;
    }
    logExceptionOnExit(toThrow);
    process['exit'](status);
  };

  Module['inspect'] = function () { return '[Emscripten Module object]'; };

} else
if (ENVIRONMENT_IS_SHELL) {

  if ((typeof process == 'object' && typeof require === 'function') || typeof window == 'object' || typeof importScripts == 'function') throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)');

  if (typeof read != 'undefined') {
    read_ = function shell_read(f) {
      return read(f);
    };
  }

  readBinary = function readBinary(f) {
    let data;
    if (typeof readbuffer == 'function') {
      return new Uint8Array(readbuffer(f));
    }
    data = read(f, 'binary');
    assert(typeof data == 'object');
    return data;
  };

  readAsync = function readAsync(f, onload, onerror) {
    setTimeout(() => onload(readBinary(f)), 0);
  };

  if (typeof scriptArgs != 'undefined') {
    arguments_ = scriptArgs;
  } else if (typeof arguments != 'undefined') {
    arguments_ = arguments;
  }

  if (typeof quit == 'function') {
    quit_ = (status, toThrow) => {
      logExceptionOnExit(toThrow);
      quit(status);
    };
  }

  if (typeof print != 'undefined') {
    // Prefer to use print/printErr where they exist, as they usually work better.
    if (typeof console == 'undefined') console = /** @type{!Console} */({});
    console.log = /** @type{!function(this:Console, ...*): undefined} */ (print);
    console.warn = console.error = /** @type{!function(this:Console, ...*): undefined} */ (typeof printErr != 'undefined' ? printErr : print);
  }

} else

// Note that this includes Node.js workers when relevant (pthreads is enabled).
// Node.js workers are detected as a combination of ENVIRONMENT_IS_WORKER and
// ENVIRONMENT_IS_NODE.
if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
  if (ENVIRONMENT_IS_WORKER) { // Check worker, not web, since window could be polyfilled
    scriptDirectory = self.location.href;
  } else if (typeof document != 'undefined' && document.currentScript) { // web
    scriptDirectory = document.currentScript.src;
  }
  // blob urls look like blob:http://site.com/etc/etc and we cannot infer anything from them.
  // otherwise, slice off the final part of the url to find the script directory.
  // if scriptDirectory does not contain a slash, lastIndexOf will return -1,
  // and scriptDirectory will correctly be replaced with an empty string.
  // If scriptDirectory contains a query (starting with ?) or a fragment (starting with #),
  // they are removed because they could contain a slash.
  if (scriptDirectory.indexOf('blob:') !== 0) {
    scriptDirectory = scriptDirectory.substr(0, scriptDirectory.replace(/[?#].*/, "").lastIndexOf('/')+1);
  } else {
    scriptDirectory = '';
  }

  if (!(typeof window == 'object' || typeof importScripts == 'function')) throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)');

  // Differentiate the Web Worker from the Node Worker case, as reading must
  // be done differently.
  {
// include: web_or_worker_shell_read.js


  read_ = (url) => {
      var xhr = new XMLHttpRequest();
      xhr.open('GET', url, false);
      xhr.send(null);
      return xhr.responseText;
  }

  if (ENVIRONMENT_IS_WORKER) {
    readBinary = (url) => {
        var xhr = new XMLHttpRequest();
        xhr.open('GET', url, false);
        xhr.responseType = 'arraybuffer';
        xhr.send(null);
        return new Uint8Array(/** @type{!ArrayBuffer} */(xhr.response));
    };
  }

  readAsync = (url, onload, onerror) => {
    var xhr = new XMLHttpRequest();
    xhr.open('GET', url, true);
    xhr.responseType = 'arraybuffer';
    xhr.onload = () => {
      if (xhr.status == 200 || (xhr.status == 0 && xhr.response)) { // file URLs can return 0
        onload(xhr.response);
        return;
      }
      onerror();
    };
    xhr.onerror = onerror;
    xhr.send(null);
  }

// end include: web_or_worker_shell_read.js
  }

  setWindowTitle = (title) => document.title = title;
} else
{
  throw new Error('environment detection error');
}

var out = Module['print'] || console.log.bind(console);
var err = Module['printErr'] || console.warn.bind(console);

// Merge back in the overrides
Object.assign(Module, moduleOverrides);
// Free the object hierarchy contained in the overrides, this lets the GC
// reclaim data used e.g. in memoryInitializerRequest, which is a large typed array.
moduleOverrides = null;
checkIncomingModuleAPI();

// Emit code to handle expected values on the Module object. This applies Module.x
// to the proper local x. This has two benefits: first, we only emit it if it is
// expected to arrive, and second, by using a local everywhere else that can be
// minified.

if (Module['arguments']) arguments_ = Module['arguments'];legacyModuleProp('arguments', 'arguments_');

if (Module['thisProgram']) thisProgram = Module['thisProgram'];legacyModuleProp('thisProgram', 'thisProgram');

if (Module['quit']) quit_ = Module['quit'];legacyModuleProp('quit', 'quit_');

// perform assertions in shell.js after we set up out() and err(), as otherwise if an assertion fails it cannot print the message
// Assertions on removed incoming Module JS APIs.
assert(typeof Module['memoryInitializerPrefixURL'] == 'undefined', 'Module.memoryInitializerPrefixURL option was removed, use Module.locateFile instead');
assert(typeof Module['pthreadMainPrefixURL'] == 'undefined', 'Module.pthreadMainPrefixURL option was removed, use Module.locateFile instead');
assert(typeof Module['cdInitializerPrefixURL'] == 'undefined', 'Module.cdInitializerPrefixURL option was removed, use Module.locateFile instead');
assert(typeof Module['filePackagePrefixURL'] == 'undefined', 'Module.filePackagePrefixURL option was removed, use Module.locateFile instead');
assert(typeof Module['read'] == 'undefined', 'Module.read option was removed (modify read_ in JS)');
assert(typeof Module['readAsync'] == 'undefined', 'Module.readAsync option was removed (modify readAsync in JS)');
assert(typeof Module['readBinary'] == 'undefined', 'Module.readBinary option was removed (modify readBinary in JS)');
assert(typeof Module['setWindowTitle'] == 'undefined', 'Module.setWindowTitle option was removed (modify setWindowTitle in JS)');
assert(typeof Module['TOTAL_MEMORY'] == 'undefined', 'Module.TOTAL_MEMORY has been renamed Module.INITIAL_MEMORY');
legacyModuleProp('read', 'read_');
legacyModuleProp('readAsync', 'readAsync');
legacyModuleProp('readBinary', 'readBinary');
legacyModuleProp('setWindowTitle', 'setWindowTitle');
var IDBFS = 'IDBFS is no longer included by default; build with -lidbfs.js';
var PROXYFS = 'PROXYFS is no longer included by default; build with -lproxyfs.js';
var WORKERFS = 'WORKERFS is no longer included by default; build with -lworkerfs.js';
var NODEFS = 'NODEFS is no longer included by default; build with -lnodefs.js';
function alignMemory() { abort('`alignMemory` is now a library function and not included by default; add it to your library.js __deps or to DEFAULT_LIBRARY_FUNCS_TO_INCLUDE on the command line'); }

assert(!ENVIRONMENT_IS_SHELL, "shell environment detected but not enabled at build time.  Add 'shell' to `-sENVIRONMENT` to enable.");




var STACK_ALIGN = 16;
var POINTER_SIZE = 4;

function getNativeTypeSize(type) {
  switch (type) {
    case 'i1': case 'i8': case 'u8': return 1;
    case 'i16': case 'u16': return 2;
    case 'i32': case 'u32': return 4;
    case 'i64': case 'u64': return 8;
    case 'float': return 4;
    case 'double': return 8;
    default: {
      if (type[type.length - 1] === '*') {
        return POINTER_SIZE;
      } else if (type[0] === 'i') {
        const bits = Number(type.substr(1));
        assert(bits % 8 === 0, 'getNativeTypeSize invalid bits ' + bits + ', type ' + type);
        return bits / 8;
      } else {
        return 0;
      }
    }
  }
}

function warnOnce(text) {
  if (!warnOnce.shown) warnOnce.shown = {};
  if (!warnOnce.shown[text]) {
    warnOnce.shown[text] = 1;
    err(text);
  }
}

// include: runtime_functions.js


// This gives correct answers for everything less than 2^{14} = 16384
// I hope nobody is contemplating functions with 16384 arguments...
function uleb128Encode(n) {
  assert(n < 16384);
  if (n < 128) {
    return [n];
  }
  return [(n % 128) | 128, n >> 7];
}

// Converts a signature like 'vii' into a description of the wasm types, like
// { parameters: ['i32', 'i32'], results: [] }.
function sigToWasmTypes(sig) {
  var typeNames = {
    'i': 'i32',
    'j': 'i64',
    'f': 'f32',
    'd': 'f64',
    'p': 'i32',
  };
  var type = {
    parameters: [],
    results: sig[0] == 'v' ? [] : [typeNames[sig[0]]]
  };
  for (var i = 1; i < sig.length; ++i) {
    assert(sig[i] in typeNames, 'invalid signature char: ' + sig[i]);
    type.parameters.push(typeNames[sig[i]]);
  }
  return type;
}

// Wraps a JS function as a wasm function with a given signature.
function convertJsFunctionToWasm(func, sig) {

  // If the type reflection proposal is available, use the new
  // "WebAssembly.Function" constructor.
  // Otherwise, construct a minimal wasm module importing the JS function and
  // re-exporting it.
  if (typeof WebAssembly.Function == "function") {
    return new WebAssembly.Function(sigToWasmTypes(sig), func);
  }

  // The module is static, with the exception of the type section, which is
  // generated based on the signature passed in.
  var typeSection = [
    0x01, // count: 1
    0x60, // form: func
  ];
  var sigRet = sig.slice(0, 1);
  var sigParam = sig.slice(1);
  var typeCodes = {
    'i': 0x7f, // i32
    'p': 0x7f, // i32
    'j': 0x7e, // i64
    'f': 0x7d, // f32
    'd': 0x7c, // f64
  };

  // Parameters, length + signatures
  typeSection = typeSection.concat(uleb128Encode(sigParam.length));
  for (var i = 0; i < sigParam.length; ++i) {
    assert(sigParam[i] in typeCodes, 'invalid signature char: ' + sigParam[i]);
    typeSection.push(typeCodes[sigParam[i]]);
  }

  // Return values, length + signatures
  // With no multi-return in MVP, either 0 (void) or 1 (anything else)
  if (sigRet == 'v') {
    typeSection.push(0x00);
  } else {
    typeSection = typeSection.concat([0x01, typeCodes[sigRet]]);
  }

  // Write the section code and overall length of the type section into the
  // section header
  typeSection = [0x01 /* Type section code */].concat(
    uleb128Encode(typeSection.length),
    typeSection
  );

  // Rest of the module is static
  var bytes = new Uint8Array([
    0x00, 0x61, 0x73, 0x6d, // magic ("\0asm")
    0x01, 0x00, 0x00, 0x00, // version: 1
  ].concat(typeSection, [
    0x02, 0x07, // import section
      // (import "e" "f" (func 0 (type 0)))
      0x01, 0x01, 0x65, 0x01, 0x66, 0x00, 0x00,
    0x07, 0x05, // export section
      // (export "f" (func 0 (type 0)))
      0x01, 0x01, 0x66, 0x00, 0x00,
  ]));

   // We can compile this wasm module synchronously because it is very small.
  // This accepts an import (at "e.f"), that it reroutes to an export (at "f")
  var module = new WebAssembly.Module(bytes);
  var instance = new WebAssembly.Instance(module, {
    'e': {
      'f': func
    }
  });
  var wrappedFunc = instance.exports['f'];
  return wrappedFunc;
}

var freeTableIndexes = [];

// Weak map of functions in the table to their indexes, created on first use.
var functionsInTableMap;

function getEmptyTableSlot() {
  // Reuse a free index if there is one, otherwise grow.
  if (freeTableIndexes.length) {
    return freeTableIndexes.pop();
  }
  // Grow the table
  try {
    wasmTable.grow(1);
  } catch (err) {
    if (!(err instanceof RangeError)) {
      throw err;
    }
    throw 'Unable to grow wasm table. Set ALLOW_TABLE_GROWTH.';
  }
  return wasmTable.length - 1;
}

function updateTableMap(offset, count) {
  for (var i = offset; i < offset + count; i++) {
    var item = getWasmTableEntry(i);
    // Ignore null values.
    if (item) {
      functionsInTableMap.set(item, i);
    }
  }
}

/**
 * Add a function to the table.
 * 'sig' parameter is required if the function being added is a JS function.
 * @param {string=} sig
 */
function addFunction(func, sig) {
  assert(typeof func != 'undefined');

  // Check if the function is already in the table, to ensure each function
  // gets a unique index. First, create the map if this is the first use.
  if (!functionsInTableMap) {
    functionsInTableMap = new WeakMap();
    updateTableMap(0, wasmTable.length);
  }
  if (functionsInTableMap.has(func)) {
    return functionsInTableMap.get(func);
  }

  // It's not in the table, add it now.

  var ret = getEmptyTableSlot();

  // Set the new value.
  try {
    // Attempting to call this with JS function will cause of table.set() to fail
    setWasmTableEntry(ret, func);
  } catch (err) {
    if (!(err instanceof TypeError)) {
      throw err;
    }
    assert(typeof sig != 'undefined', 'Missing signature argument to addFunction: ' + func);
    var wrapped = convertJsFunctionToWasm(func, sig);
    setWasmTableEntry(ret, wrapped);
  }

  functionsInTableMap.set(func, ret);

  return ret;
}

function removeFunction(index) {
  functionsInTableMap.delete(getWasmTableEntry(index));
  freeTableIndexes.push(index);
}

// end include: runtime_functions.js
// include: runtime_debug.js


function legacyModuleProp(prop, newName) {
  if (!Object.getOwnPropertyDescriptor(Module, prop)) {
    Object.defineProperty(Module, prop, {
      configurable: true,
      get: function() {
        abort('Module.' + prop + ' has been replaced with plain ' + newName + ' (the initial value can be provided on Module, but after startup the value is only looked for on a local variable of that name)');
      }
    });
  }
}

function ignoredModuleProp(prop) {
  if (Object.getOwnPropertyDescriptor(Module, prop)) {
    abort('`Module.' + prop + '` was supplied but `' + prop + '` not included in INCOMING_MODULE_JS_API');
  }
}

function unexportedMessage(sym, isFSSybol) {
  var msg = "'" + sym + "' was not exported. add it to EXPORTED_RUNTIME_METHODS (see the FAQ)";
  if (isFSSybol) {
    msg += '. Alternatively, forcing filesystem support (-sFORCE_FILESYSTEM) can export this for you';
  }
  return msg;
}

function unexportedRuntimeSymbol(sym, isFSSybol) {
  if (!Object.getOwnPropertyDescriptor(Module, sym)) {
    Object.defineProperty(Module, sym, {
      configurable: true,
      get: function() {
        abort(unexportedMessage(sym, isFSSybol));
      }
    });
  }
}

function unexportedRuntimeFunction(sym, isFSSybol) {
  if (!Object.getOwnPropertyDescriptor(Module, sym)) {
    Module[sym] = () => abort(unexportedMessage(sym, isFSSybol));
  }
}

// end include: runtime_debug.js
var tempRet0 = 0;
var setTempRet0 = (value) => { tempRet0 = value; };
var getTempRet0 = () => tempRet0;



// === Preamble library stuff ===

// Documentation for the public APIs defined in this file must be updated in:
//    site/source/docs/api_reference/preamble.js.rst
// A prebuilt local version of the documentation is available at:
//    site/build/text/docs/api_reference/preamble.js.txt
// You can also build docs locally as HTML or other formats in site/
// An online HTML version (which may be of a different version of Emscripten)
//    is up at http://kripken.github.io/emscripten-site/docs/api_reference/preamble.js.html

var wasmBinary;
if (Module['wasmBinary']) wasmBinary = Module['wasmBinary'];legacyModuleProp('wasmBinary', 'wasmBinary');
var noExitRuntime = Module['noExitRuntime'] || true;legacyModuleProp('noExitRuntime', 'noExitRuntime');

if (typeof WebAssembly != 'object') {
  abort('no native wasm support detected');
}

// Wasm globals

var wasmMemory;

//========================================
// Runtime essentials
//========================================

// whether we are quitting the application. no code should run after this.
// set in exit() and abort()
var ABORT = false;

// set by exit() and abort().  Passed to 'onExit' handler.
// NOTE: This is also used as the process return code code in shell environments
// but only when noExitRuntime is false.
var EXITSTATUS;

/** @type {function(*, string=)} */
function assert(condition, text) {
  if (!condition) {
    abort('Assertion failed' + (text ? ': ' + text : ''));
  }
}

// Returns the C function with a specified identifier (for C++, you need to do manual name mangling)
function getCFunc(ident) {
  var func = Module['_' + ident]; // closure exported function
  assert(func, 'Cannot call unknown function ' + ident + ', make sure it is exported');
  return func;
}

// C calling interface.
/** @param {string|null=} returnType
    @param {Array=} argTypes
    @param {Arguments|Array=} args
    @param {Object=} opts */
function ccall(ident, returnType, argTypes, args, opts) {
  // For fast lookup of conversion functions
  var toC = {
    'string': function(str) {
      var ret = 0;
      if (str !== null && str !== undefined && str !== 0) { // null string
        // at most 4 bytes per UTF-8 code point, +1 for the trailing '\0'
        var len = (str.length << 2) + 1;
        ret = stackAlloc(len);
        stringToUTF8(str, ret, len);
      }
      return ret;
    },
    'array': function(arr) {
      var ret = stackAlloc(arr.length);
      writeArrayToMemory(arr, ret);
      return ret;
    }
  };

  function convertReturnValue(ret) {
    if (returnType === 'string') {
      
      return UTF8ToString(ret);
    }
    if (returnType === 'boolean') return Boolean(ret);
    return ret;
  }

  var func = getCFunc(ident);
  var cArgs = [];
  var stack = 0;
  assert(returnType !== 'array', 'Return type should not be "array".');
  if (args) {
    for (var i = 0; i < args.length; i++) {
      var converter = toC[argTypes[i]];
      if (converter) {
        if (stack === 0) stack = stackSave();
        cArgs[i] = converter(args[i]);
      } else {
        cArgs[i] = args[i];
      }
    }
  }
  var ret = func.apply(null, cArgs);
  function onDone(ret) {
    if (stack !== 0) stackRestore(stack);
    return convertReturnValue(ret);
  }

  ret = onDone(ret);
  return ret;
}

/** @param {string=} returnType
    @param {Array=} argTypes
    @param {Object=} opts */
function cwrap(ident, returnType, argTypes, opts) {
  return function() {
    return ccall(ident, returnType, argTypes, arguments, opts);
  }
}

// We used to include malloc/free by default in the past. Show a helpful error in
// builds with assertions.
function _free() {
  // Show a helpful error since we used to include free by default in the past.
  abort("free() called but not included in the build - add '_free' to EXPORTED_FUNCTIONS");
}

// include: runtime_legacy.js


var ALLOC_NORMAL = 0; // Tries to use _malloc()
var ALLOC_STACK = 1; // Lives for the duration of the current function call

/**
 * allocate(): This function is no longer used by emscripten but is kept around to avoid
 *             breaking external users.
 *             You should normally not use allocate(), and instead allocate
 *             memory using _malloc()/stackAlloc(), initialize it with
 *             setValue(), and so forth.
 * @param {(Uint8Array|Array<number>)} slab: An array of data.
 * @param {number=} allocator : How to allocate memory, see ALLOC_*
 */
function allocate(slab, allocator) {
  var ret;
  assert(typeof allocator == 'number', 'allocate no longer takes a type argument')
  assert(typeof slab != 'number', 'allocate no longer takes a number as arg0')

  if (allocator == ALLOC_STACK) {
    ret = stackAlloc(slab.length);
  } else {
    ret = _malloc(slab.length);
  }

  if (!slab.subarray && !slab.slice) {
    slab = new Uint8Array(slab);
  }
  HEAPU8.set(slab, ret);
  return ret;
}

// end include: runtime_legacy.js
// include: runtime_strings.js


// runtime_strings.js: Strings related runtime functions that are part of both MINIMAL_RUNTIME and regular runtime.

var UTF8Decoder = typeof TextDecoder != 'undefined' ? new TextDecoder('utf8') : undefined;

// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
// a copy of that string as a Javascript String object.
/**
 * heapOrArray is either a regular array, or a JavaScript typed array view.
 * @param {number} idx
 * @param {number=} maxBytesToRead
 * @return {string}
 */
function UTF8ArrayToString(heapOrArray, idx, maxBytesToRead) {
  var endIdx = idx + maxBytesToRead;
  var endPtr = idx;
  // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
  // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
  // (As a tiny code save trick, compare endPtr against endIdx using a negation, so that undefined means Infinity)
  while (heapOrArray[endPtr] && !(endPtr >= endIdx)) ++endPtr;

  if (endPtr - idx > 16 && heapOrArray.buffer && UTF8Decoder) {
    return UTF8Decoder.decode(heapOrArray.subarray(idx, endPtr));
  } else {
    var str = '';
    // If building with TextDecoder, we have already computed the string length above, so test loop end condition against that
    while (idx < endPtr) {
      // For UTF8 byte structure, see:
      // http://en.wikipedia.org/wiki/UTF-8#Description
      // https://www.ietf.org/rfc/rfc2279.txt
      // https://tools.ietf.org/html/rfc3629
      var u0 = heapOrArray[idx++];
      if (!(u0 & 0x80)) { str += String.fromCharCode(u0); continue; }
      var u1 = heapOrArray[idx++] & 63;
      if ((u0 & 0xE0) == 0xC0) { str += String.fromCharCode(((u0 & 31) << 6) | u1); continue; }
      var u2 = heapOrArray[idx++] & 63;
      if ((u0 & 0xF0) == 0xE0) {
        u0 = ((u0 & 15) << 12) | (u1 << 6) | u2;
      } else {
        if ((u0 & 0xF8) != 0xF0) warnOnce('Invalid UTF-8 leading byte 0x' + u0.toString(16) + ' encountered when deserializing a UTF-8 string in wasm memory to a JS string!');
        u0 = ((u0 & 7) << 18) | (u1 << 12) | (u2 << 6) | (heapOrArray[idx++] & 63);
      }

      if (u0 < 0x10000) {
        str += String.fromCharCode(u0);
      } else {
        var ch = u0 - 0x10000;
        str += String.fromCharCode(0xD800 | (ch >> 10), 0xDC00 | (ch & 0x3FF));
      }
    }
  }
  return str;
}

// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the emscripten HEAP, returns a
// copy of that string as a Javascript String object.
// maxBytesToRead: an optional length that specifies the maximum number of bytes to read. You can omit
//                 this parameter to scan the string until the first \0 byte. If maxBytesToRead is
//                 passed, and the string at [ptr, ptr+maxBytesToReadr[ contains a null byte in the
//                 middle, then the string will cut short at that byte index (i.e. maxBytesToRead will
//                 not produce a string of exact length [ptr, ptr+maxBytesToRead[)
//                 N.B. mixing frequent uses of UTF8ToString() with and without maxBytesToRead may
//                 throw JS JIT optimizations off, so it is worth to consider consistently using one
//                 style or the other.
/**
 * @param {number} ptr
 * @param {number=} maxBytesToRead
 * @return {string}
 */
function UTF8ToString(ptr, maxBytesToRead) {
  return ptr ? UTF8ArrayToString(HEAPU8, ptr, maxBytesToRead) : '';
}

// Copies the given Javascript String object 'str' to the given byte array at address 'outIdx',
// encoded in UTF8 form and null-terminated. The copy will require at most str.length*4+1 bytes of space in the HEAP.
// Use the function lengthBytesUTF8 to compute the exact number of bytes (excluding null terminator) that this function will write.
// Parameters:
//   str: the Javascript string to copy.
//   heap: the array to copy to. Each index in this array is assumed to be one 8-byte element.
//   outIdx: The starting offset in the array to begin the copying.
//   maxBytesToWrite: The maximum number of bytes this function can write to the array.
//                    This count should include the null terminator,
//                    i.e. if maxBytesToWrite=1, only the null terminator will be written and nothing else.
//                    maxBytesToWrite=0 does not write any bytes to the output, not even the null terminator.
// Returns the number of bytes written, EXCLUDING the null terminator.

function stringToUTF8Array(str, heap, outIdx, maxBytesToWrite) {
  if (!(maxBytesToWrite > 0)) // Parameter maxBytesToWrite is not optional. Negative values, 0, null, undefined and false each don't write out any bytes.
    return 0;

  var startIdx = outIdx;
  var endIdx = outIdx + maxBytesToWrite - 1; // -1 for string null terminator.
  for (var i = 0; i < str.length; ++i) {
    // Gotcha: charCodeAt returns a 16-bit word that is a UTF-16 encoded code unit, not a Unicode code point of the character! So decode UTF16->UTF32->UTF8.
    // See http://unicode.org/faq/utf_bom.html#utf16-3
    // For UTF8 byte structure, see http://en.wikipedia.org/wiki/UTF-8#Description and https://www.ietf.org/rfc/rfc2279.txt and https://tools.ietf.org/html/rfc3629
    var u = str.charCodeAt(i); // possibly a lead surrogate
    if (u >= 0xD800 && u <= 0xDFFF) {
      var u1 = str.charCodeAt(++i);
      u = 0x10000 + ((u & 0x3FF) << 10) | (u1 & 0x3FF);
    }
    if (u <= 0x7F) {
      if (outIdx >= endIdx) break;
      heap[outIdx++] = u;
    } else if (u <= 0x7FF) {
      if (outIdx + 1 >= endIdx) break;
      heap[outIdx++] = 0xC0 | (u >> 6);
      heap[outIdx++] = 0x80 | (u & 63);
    } else if (u <= 0xFFFF) {
      if (outIdx + 2 >= endIdx) break;
      heap[outIdx++] = 0xE0 | (u >> 12);
      heap[outIdx++] = 0x80 | ((u >> 6) & 63);
      heap[outIdx++] = 0x80 | (u & 63);
    } else {
      if (outIdx + 3 >= endIdx) break;
      if (u > 0x10FFFF) warnOnce('Invalid Unicode code point 0x' + u.toString(16) + ' encountered when serializing a JS string to a UTF-8 string in wasm memory! (Valid unicode code points should be in range 0-0x10FFFF).');
      heap[outIdx++] = 0xF0 | (u >> 18);
      heap[outIdx++] = 0x80 | ((u >> 12) & 63);
      heap[outIdx++] = 0x80 | ((u >> 6) & 63);
      heap[outIdx++] = 0x80 | (u & 63);
    }
  }
  // Null-terminate the pointer to the buffer.
  heap[outIdx] = 0;
  return outIdx - startIdx;
}

// Copies the given Javascript String object 'str' to the emscripten HEAP at address 'outPtr',
// null-terminated and encoded in UTF8 form. The copy will require at most str.length*4+1 bytes of space in the HEAP.
// Use the function lengthBytesUTF8 to compute the exact number of bytes (excluding null terminator) that this function will write.
// Returns the number of bytes written, EXCLUDING the null terminator.

function stringToUTF8(str, outPtr, maxBytesToWrite) {
  assert(typeof maxBytesToWrite == 'number', 'stringToUTF8(str, outPtr, maxBytesToWrite) is missing the third parameter that specifies the length of the output buffer!');
  return stringToUTF8Array(str, HEAPU8,outPtr, maxBytesToWrite);
}

// Returns the number of bytes the given Javascript string takes if encoded as a UTF8 byte array, EXCLUDING the null terminator byte.
function lengthBytesUTF8(str) {
  var len = 0;
  for (var i = 0; i < str.length; ++i) {
    // Gotcha: charCodeAt returns a 16-bit word that is a UTF-16 encoded code unit, not a Unicode code point of the character! So decode UTF16->UTF32->UTF8.
    // See http://unicode.org/faq/utf_bom.html#utf16-3
    var u = str.charCodeAt(i); // possibly a lead surrogate
    if (u >= 0xD800 && u <= 0xDFFF) u = 0x10000 + ((u & 0x3FF) << 10) | (str.charCodeAt(++i) & 0x3FF);
    if (u <= 0x7F) ++len;
    else if (u <= 0x7FF) len += 2;
    else if (u <= 0xFFFF) len += 3;
    else len += 4;
  }
  return len;
}

// end include: runtime_strings.js
// include: runtime_strings_extra.js


// runtime_strings_extra.js: Strings related runtime functions that are available only in regular runtime.

// Given a pointer 'ptr' to a null-terminated ASCII-encoded string in the emscripten HEAP, returns
// a copy of that string as a Javascript String object.

function AsciiToString(ptr) {
  var str = '';
  while (1) {
    var ch = HEAPU8[((ptr++)>>0)];
    if (!ch) return str;
    str += String.fromCharCode(ch);
  }
}

// Copies the given Javascript String object 'str' to the emscripten HEAP at address 'outPtr',
// null-terminated and encoded in ASCII form. The copy will require at most str.length+1 bytes of space in the HEAP.

function stringToAscii(str, outPtr) {
  return writeAsciiToMemory(str, outPtr, false);
}

// Given a pointer 'ptr' to a null-terminated UTF16LE-encoded string in the emscripten HEAP, returns
// a copy of that string as a Javascript String object.

var UTF16Decoder = typeof TextDecoder != 'undefined' ? new TextDecoder('utf-16le') : undefined;

function UTF16ToString(ptr, maxBytesToRead) {
  assert(ptr % 2 == 0, 'Pointer passed to UTF16ToString must be aligned to two bytes!');
  var endPtr = ptr;
  // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
  // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
  var idx = endPtr >> 1;
  var maxIdx = idx + maxBytesToRead / 2;
  // If maxBytesToRead is not passed explicitly, it will be undefined, and this
  // will always evaluate to true. This saves on code size.
  while (!(idx >= maxIdx) && HEAPU16[idx]) ++idx;
  endPtr = idx << 1;

  if (endPtr - ptr > 32 && UTF16Decoder) {
    return UTF16Decoder.decode(HEAPU8.subarray(ptr, endPtr));
  } else {
    var str = '';

    // If maxBytesToRead is not passed explicitly, it will be undefined, and the for-loop's condition
    // will always evaluate to true. The loop is then terminated on the first null char.
    for (var i = 0; !(i >= maxBytesToRead / 2); ++i) {
      var codeUnit = HEAP16[(((ptr)+(i*2))>>1)];
      if (codeUnit == 0) break;
      // fromCharCode constructs a character from a UTF-16 code unit, so we can pass the UTF16 string right through.
      str += String.fromCharCode(codeUnit);
    }

    return str;
  }
}

// Copies the given Javascript String object 'str' to the emscripten HEAP at address 'outPtr',
// null-terminated and encoded in UTF16 form. The copy will require at most str.length*4+2 bytes of space in the HEAP.
// Use the function lengthBytesUTF16() to compute the exact number of bytes (excluding null terminator) that this function will write.
// Parameters:
//   str: the Javascript string to copy.
//   outPtr: Byte address in Emscripten HEAP where to write the string to.
//   maxBytesToWrite: The maximum number of bytes this function can write to the array. This count should include the null
//                    terminator, i.e. if maxBytesToWrite=2, only the null terminator will be written and nothing else.
//                    maxBytesToWrite<2 does not write any bytes to the output, not even the null terminator.
// Returns the number of bytes written, EXCLUDING the null terminator.

function stringToUTF16(str, outPtr, maxBytesToWrite) {
  assert(outPtr % 2 == 0, 'Pointer passed to stringToUTF16 must be aligned to two bytes!');
  assert(typeof maxBytesToWrite == 'number', 'stringToUTF16(str, outPtr, maxBytesToWrite) is missing the third parameter that specifies the length of the output buffer!');
  // Backwards compatibility: if max bytes is not specified, assume unsafe unbounded write is allowed.
  if (maxBytesToWrite === undefined) {
    maxBytesToWrite = 0x7FFFFFFF;
  }
  if (maxBytesToWrite < 2) return 0;
  maxBytesToWrite -= 2; // Null terminator.
  var startPtr = outPtr;
  var numCharsToWrite = (maxBytesToWrite < str.length*2) ? (maxBytesToWrite / 2) : str.length;
  for (var i = 0; i < numCharsToWrite; ++i) {
    // charCodeAt returns a UTF-16 encoded code unit, so it can be directly written to the HEAP.
    var codeUnit = str.charCodeAt(i); // possibly a lead surrogate
    HEAP16[((outPtr)>>1)] = codeUnit;
    outPtr += 2;
  }
  // Null-terminate the pointer to the HEAP.
  HEAP16[((outPtr)>>1)] = 0;
  return outPtr - startPtr;
}

// Returns the number of bytes the given Javascript string takes if encoded as a UTF16 byte array, EXCLUDING the null terminator byte.

function lengthBytesUTF16(str) {
  return str.length*2;
}

function UTF32ToString(ptr, maxBytesToRead) {
  assert(ptr % 4 == 0, 'Pointer passed to UTF32ToString must be aligned to four bytes!');
  var i = 0;

  var str = '';
  // If maxBytesToRead is not passed explicitly, it will be undefined, and this
  // will always evaluate to true. This saves on code size.
  while (!(i >= maxBytesToRead / 4)) {
    var utf32 = HEAP32[(((ptr)+(i*4))>>2)];
    if (utf32 == 0) break;
    ++i;
    // Gotcha: fromCharCode constructs a character from a UTF-16 encoded code (pair), not from a Unicode code point! So encode the code point to UTF-16 for constructing.
    // See http://unicode.org/faq/utf_bom.html#utf16-3
    if (utf32 >= 0x10000) {
      var ch = utf32 - 0x10000;
      str += String.fromCharCode(0xD800 | (ch >> 10), 0xDC00 | (ch & 0x3FF));
    } else {
      str += String.fromCharCode(utf32);
    }
  }
  return str;
}

// Copies the given Javascript String object 'str' to the emscripten HEAP at address 'outPtr',
// null-terminated and encoded in UTF32 form. The copy will require at most str.length*4+4 bytes of space in the HEAP.
// Use the function lengthBytesUTF32() to compute the exact number of bytes (excluding null terminator) that this function will write.
// Parameters:
//   str: the Javascript string to copy.
//   outPtr: Byte address in Emscripten HEAP where to write the string to.
//   maxBytesToWrite: The maximum number of bytes this function can write to the array. This count should include the null
//                    terminator, i.e. if maxBytesToWrite=4, only the null terminator will be written and nothing else.
//                    maxBytesToWrite<4 does not write any bytes to the output, not even the null terminator.
// Returns the number of bytes written, EXCLUDING the null terminator.

function stringToUTF32(str, outPtr, maxBytesToWrite) {
  assert(outPtr % 4 == 0, 'Pointer passed to stringToUTF32 must be aligned to four bytes!');
  assert(typeof maxBytesToWrite == 'number', 'stringToUTF32(str, outPtr, maxBytesToWrite) is missing the third parameter that specifies the length of the output buffer!');
  // Backwards compatibility: if max bytes is not specified, assume unsafe unbounded write is allowed.
  if (maxBytesToWrite === undefined) {
    maxBytesToWrite = 0x7FFFFFFF;
  }
  if (maxBytesToWrite < 4) return 0;
  var startPtr = outPtr;
  var endPtr = startPtr + maxBytesToWrite - 4;
  for (var i = 0; i < str.length; ++i) {
    // Gotcha: charCodeAt returns a 16-bit word that is a UTF-16 encoded code unit, not a Unicode code point of the character! We must decode the string to UTF-32 to the heap.
    // See http://unicode.org/faq/utf_bom.html#utf16-3
    var codeUnit = str.charCodeAt(i); // possibly a lead surrogate
    if (codeUnit >= 0xD800 && codeUnit <= 0xDFFF) {
      var trailSurrogate = str.charCodeAt(++i);
      codeUnit = 0x10000 + ((codeUnit & 0x3FF) << 10) | (trailSurrogate & 0x3FF);
    }
    HEAP32[((outPtr)>>2)] = codeUnit;
    outPtr += 4;
    if (outPtr + 4 > endPtr) break;
  }
  // Null-terminate the pointer to the HEAP.
  HEAP32[((outPtr)>>2)] = 0;
  return outPtr - startPtr;
}

// Returns the number of bytes the given Javascript string takes if encoded as a UTF16 byte array, EXCLUDING the null terminator byte.

function lengthBytesUTF32(str) {
  var len = 0;
  for (var i = 0; i < str.length; ++i) {
    // Gotcha: charCodeAt returns a 16-bit word that is a UTF-16 encoded code unit, not a Unicode code point of the character! We must decode the string to UTF-32 to the heap.
    // See http://unicode.org/faq/utf_bom.html#utf16-3
    var codeUnit = str.charCodeAt(i);
    if (codeUnit >= 0xD800 && codeUnit <= 0xDFFF) ++i; // possibly a lead surrogate, so skip over the tail surrogate.
    len += 4;
  }

  return len;
}

// Allocate heap space for a JS string, and write it there.
// It is the responsibility of the caller to free() that memory.
function allocateUTF8(str) {
  var size = lengthBytesUTF8(str) + 1;
  var ret = _malloc(size);
  if (ret) stringToUTF8Array(str, HEAP8, ret, size);
  return ret;
}

// Allocate stack space for a JS string, and write it there.
function allocateUTF8OnStack(str) {
  var size = lengthBytesUTF8(str) + 1;
  var ret = stackAlloc(size);
  stringToUTF8Array(str, HEAP8, ret, size);
  return ret;
}

// Deprecated: This function should not be called because it is unsafe and does not provide
// a maximum length limit of how many bytes it is allowed to write. Prefer calling the
// function stringToUTF8Array() instead, which takes in a maximum length that can be used
// to be secure from out of bounds writes.
/** @deprecated
    @param {boolean=} dontAddNull */
function writeStringToMemory(string, buffer, dontAddNull) {
  warnOnce('writeStringToMemory is deprecated and should not be called! Use stringToUTF8() instead!');

  var /** @type {number} */ lastChar, /** @type {number} */ end;
  if (dontAddNull) {
    // stringToUTF8Array always appends null. If we don't want to do that, remember the
    // character that existed at the location where the null will be placed, and restore
    // that after the write (below).
    end = buffer + lengthBytesUTF8(string);
    lastChar = HEAP8[end];
  }
  stringToUTF8(string, buffer, Infinity);
  if (dontAddNull) HEAP8[end] = lastChar; // Restore the value under the null character.
}

function writeArrayToMemory(array, buffer) {
  assert(array.length >= 0, 'writeArrayToMemory array must have a length (should be an array or typed array)')
  HEAP8.set(array, buffer);
}

/** @param {boolean=} dontAddNull */
function writeAsciiToMemory(str, buffer, dontAddNull) {
  for (var i = 0; i < str.length; ++i) {
    assert(str.charCodeAt(i) === (str.charCodeAt(i) & 0xff));
    HEAP8[((buffer++)>>0)] = str.charCodeAt(i);
  }
  // Null-terminate the pointer to the HEAP.
  if (!dontAddNull) HEAP8[((buffer)>>0)] = 0;
}

// end include: runtime_strings_extra.js
// Memory management

var HEAP,
/** @type {!ArrayBuffer} */
  buffer,
/** @type {!Int8Array} */
  HEAP8,
/** @type {!Uint8Array} */
  HEAPU8,
/** @type {!Int16Array} */
  HEAP16,
/** @type {!Uint16Array} */
  HEAPU16,
/** @type {!Int32Array} */
  HEAP32,
/** @type {!Uint32Array} */
  HEAPU32,
/** @type {!Float32Array} */
  HEAPF32,
/** @type {!Float64Array} */
  HEAPF64;

function updateGlobalBufferAndViews(buf) {
  buffer = buf;
  Module['HEAP8'] = HEAP8 = new Int8Array(buf);
  Module['HEAP16'] = HEAP16 = new Int16Array(buf);
  Module['HEAP32'] = HEAP32 = new Int32Array(buf);
  Module['HEAPU8'] = HEAPU8 = new Uint8Array(buf);
  Module['HEAPU16'] = HEAPU16 = new Uint16Array(buf);
  Module['HEAPU32'] = HEAPU32 = new Uint32Array(buf);
  Module['HEAPF32'] = HEAPF32 = new Float32Array(buf);
  Module['HEAPF64'] = HEAPF64 = new Float64Array(buf);
}

var TOTAL_STACK = 5242880;
if (Module['TOTAL_STACK']) assert(TOTAL_STACK === Module['TOTAL_STACK'], 'the stack size can no longer be determined at runtime')

var INITIAL_MEMORY = Module['INITIAL_MEMORY'] || 16777216;legacyModuleProp('INITIAL_MEMORY', 'INITIAL_MEMORY');

assert(INITIAL_MEMORY >= TOTAL_STACK, 'INITIAL_MEMORY should be larger than TOTAL_STACK, was ' + INITIAL_MEMORY + '! (TOTAL_STACK=' + TOTAL_STACK + ')');

// check for full engine support (use string 'subarray' to avoid closure compiler confusion)
assert(typeof Int32Array != 'undefined' && typeof Float64Array !== 'undefined' && Int32Array.prototype.subarray != undefined && Int32Array.prototype.set != undefined,
       'JS engine does not provide full typed array support');

// If memory is defined in wasm, the user can't provide it.
assert(!Module['wasmMemory'], 'Use of `wasmMemory` detected.  Use -sIMPORTED_MEMORY to define wasmMemory externally');
assert(INITIAL_MEMORY == 16777216, 'Detected runtime INITIAL_MEMORY setting.  Use -sIMPORTED_MEMORY to define wasmMemory dynamically');

// include: runtime_init_table.js
// In regular non-RELOCATABLE mode the table is exported
// from the wasm module and this will be assigned once
// the exports are available.
var wasmTable;

// end include: runtime_init_table.js
// include: runtime_stack_check.js


// Initializes the stack cookie. Called at the startup of main and at the startup of each thread in pthreads mode.
function writeStackCookie() {
  var max = _emscripten_stack_get_end();
  assert((max & 3) == 0);
  // The stack grow downwards towards _emscripten_stack_get_end.
  // We write cookies to the final two words in the stack and detect if they are
  // ever overwritten.
  HEAP32[((max)>>2)] = 0x2135467;
  HEAP32[(((max)+(4))>>2)] = 0x89BACDFE;
  // Also test the global address 0 for integrity.
  HEAPU32[0] = 0x63736d65; /* 'emsc' */
}

function checkStackCookie() {
  if (ABORT) return;
  var max = _emscripten_stack_get_end();
  var cookie1 = HEAPU32[((max)>>2)];
  var cookie2 = HEAPU32[(((max)+(4))>>2)];
  if (cookie1 != 0x2135467 || cookie2 != 0x89BACDFE) {
    abort('Stack overflow! Stack cookie has been overwritten, expected hex dwords 0x89BACDFE and 0x2135467, but received 0x' + cookie2.toString(16) + ' 0x' + cookie1.toString(16));
  }
  // Also test the global address 0 for integrity.
  if (HEAPU32[0] !== 0x63736d65 /* 'emsc' */) abort('Runtime error: The application has corrupted its heap memory area (address zero)!');
}

// end include: runtime_stack_check.js
// include: runtime_assertions.js


// Endianness check
(function() {
  var h16 = new Int16Array(1);
  var h8 = new Int8Array(h16.buffer);
  h16[0] = 0x6373;
  if (h8[0] !== 0x73 || h8[1] !== 0x63) throw 'Runtime error: expected the system to be little-endian! (Run with -sSUPPORT_BIG_ENDIAN to bypass)';
})();

// end include: runtime_assertions.js
var __ATPRERUN__  = []; // functions called before the runtime is initialized
var __ATINIT__    = []; // functions called during startup
var __ATEXIT__    = []; // functions called during shutdown
var __ATPOSTRUN__ = []; // functions called after the main() is called

var runtimeInitialized = false;

function keepRuntimeAlive() {
  return noExitRuntime;
}

function preRun() {

  if (Module['preRun']) {
    if (typeof Module['preRun'] == 'function') Module['preRun'] = [Module['preRun']];
    while (Module['preRun'].length) {
      addOnPreRun(Module['preRun'].shift());
    }
  }

  callRuntimeCallbacks(__ATPRERUN__);
}

function initRuntime() {
  assert(!runtimeInitialized);
  runtimeInitialized = true;

  checkStackCookie();

  
  callRuntimeCallbacks(__ATINIT__);
}

function postRun() {
  checkStackCookie();

  if (Module['postRun']) {
    if (typeof Module['postRun'] == 'function') Module['postRun'] = [Module['postRun']];
    while (Module['postRun'].length) {
      addOnPostRun(Module['postRun'].shift());
    }
  }

  callRuntimeCallbacks(__ATPOSTRUN__);
}

function addOnPreRun(cb) {
  __ATPRERUN__.unshift(cb);
}

function addOnInit(cb) {
  __ATINIT__.unshift(cb);
}

function addOnExit(cb) {
}

function addOnPostRun(cb) {
  __ATPOSTRUN__.unshift(cb);
}

// include: runtime_math.js


// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/imul

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/fround

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/clz32

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/trunc

assert(Math.imul, 'This browser does not support Math.imul(), build with LEGACY_VM_SUPPORT or POLYFILL_OLD_MATH_FUNCTIONS to add in a polyfill');
assert(Math.fround, 'This browser does not support Math.fround(), build with LEGACY_VM_SUPPORT or POLYFILL_OLD_MATH_FUNCTIONS to add in a polyfill');
assert(Math.clz32, 'This browser does not support Math.clz32(), build with LEGACY_VM_SUPPORT or POLYFILL_OLD_MATH_FUNCTIONS to add in a polyfill');
assert(Math.trunc, 'This browser does not support Math.trunc(), build with LEGACY_VM_SUPPORT or POLYFILL_OLD_MATH_FUNCTIONS to add in a polyfill');

// end include: runtime_math.js
// A counter of dependencies for calling run(). If we need to
// do asynchronous work before running, increment this and
// decrement it. Incrementing must happen in a place like
// Module.preRun (used by emcc to add file preloading).
// Note that you can add dependencies in preRun, even though
// it happens right before run - run will be postponed until
// the dependencies are met.
var runDependencies = 0;
var runDependencyWatcher = null;
var dependenciesFulfilled = null; // overridden to take different actions when all run dependencies are fulfilled
var runDependencyTracking = {};

function getUniqueRunDependency(id) {
  var orig = id;
  while (1) {
    if (!runDependencyTracking[id]) return id;
    id = orig + Math.random();
  }
}

function addRunDependency(id) {
  runDependencies++;

  if (Module['monitorRunDependencies']) {
    Module['monitorRunDependencies'](runDependencies);
  }

  if (id) {
    assert(!runDependencyTracking[id]);
    runDependencyTracking[id] = 1;
    if (runDependencyWatcher === null && typeof setInterval != 'undefined') {
      // Check for missing dependencies every few seconds
      runDependencyWatcher = setInterval(function() {
        if (ABORT) {
          clearInterval(runDependencyWatcher);
          runDependencyWatcher = null;
          return;
        }
        var shown = false;
        for (var dep in runDependencyTracking) {
          if (!shown) {
            shown = true;
            err('still waiting on run dependencies:');
          }
          err('dependency: ' + dep);
        }
        if (shown) {
          err('(end of list)');
        }
      }, 10000);
    }
  } else {
    err('warning: run dependency added without ID');
  }
}

function removeRunDependency(id) {
  runDependencies--;

  if (Module['monitorRunDependencies']) {
    Module['monitorRunDependencies'](runDependencies);
  }

  if (id) {
    assert(runDependencyTracking[id]);
    delete runDependencyTracking[id];
  } else {
    err('warning: run dependency removed without ID');
  }
  if (runDependencies == 0) {
    if (runDependencyWatcher !== null) {
      clearInterval(runDependencyWatcher);
      runDependencyWatcher = null;
    }
    if (dependenciesFulfilled) {
      var callback = dependenciesFulfilled;
      dependenciesFulfilled = null;
      callback(); // can add another dependenciesFulfilled
    }
  }
}

/** @param {string|number=} what */
function abort(what) {
  {
    if (Module['onAbort']) {
      Module['onAbort'](what);
    }
  }

  what = 'Aborted(' + what + ')';
  // TODO(sbc): Should we remove printing and leave it up to whoever
  // catches the exception?
  err(what);

  ABORT = true;
  EXITSTATUS = 1;

  // Use a wasm runtime error, because a JS error might be seen as a foreign
  // exception, which means we'd run destructors on it. We need the error to
  // simply make the program stop.
  // FIXME This approach does not work in Wasm EH because it currently does not assume
  // all RuntimeErrors are from traps; it decides whether a RuntimeError is from
  // a trap or not based on a hidden field within the object. So at the moment
  // we don't have a way of throwing a wasm trap from JS. TODO Make a JS API that
  // allows this in the wasm spec.

  // Suppress closure compiler warning here. Closure compiler's builtin extern
  // defintion for WebAssembly.RuntimeError claims it takes no arguments even
  // though it can.
  // TODO(https://github.com/google/closure-compiler/pull/3913): Remove if/when upstream closure gets fixed.
  /** @suppress {checkTypes} */
  var e = new WebAssembly.RuntimeError(what);

  // Throw the error whether or not MODULARIZE is set because abort is used
  // in code paths apart from instantiation where an exception is expected
  // to be thrown when abort is called.
  throw e;
}

// {{MEM_INITIALIZER}}

// include: memoryprofiler.js


// end include: memoryprofiler.js
// show errors on likely calls to FS when it was not included
var FS = {
  error: function() {
    abort('Filesystem support (FS) was not included. The problem is that you are using files from JS, but files were not used from C/C++, so filesystem support was not auto-included. You can force-include filesystem support with -sFORCE_FILESYSTEM');
  },
  init: function() { FS.error() },
  createDataFile: function() { FS.error() },
  createPreloadedFile: function() { FS.error() },
  createLazyFile: function() { FS.error() },
  open: function() { FS.error() },
  mkdev: function() { FS.error() },
  registerDevice: function() { FS.error() },
  analyzePath: function() { FS.error() },
  loadFilesFromDB: function() { FS.error() },

  ErrnoError: function ErrnoError() { FS.error() },
};
Module['FS_createDataFile'] = FS.createDataFile;
Module['FS_createPreloadedFile'] = FS.createPreloadedFile;

// include: URIUtils.js


// Prefix of data URIs emitted by SINGLE_FILE and related options.
var dataURIPrefix = 'data:application/octet-stream;base64,';

// Indicates whether filename is a base64 data URI.
function isDataURI(filename) {
  // Prefix of data URIs emitted by SINGLE_FILE and related options.
  return filename.startsWith(dataURIPrefix);
}

// Indicates whether filename is delivered via file protocol (as opposed to http/https)
function isFileURI(filename) {
  return filename.startsWith('file://');
}

// end include: URIUtils.js
/** @param {boolean=} fixedasm */
function createExportWrapper(name, fixedasm) {
  return function() {
    var displayName = name;
    var asm = fixedasm;
    if (!fixedasm) {
      asm = Module['asm'];
    }
    assert(runtimeInitialized, 'native function `' + displayName + '` called before runtime initialization');
    if (!asm[name]) {
      assert(asm[name], 'exported native function `' + displayName + '` not found');
    }
    return asm[name].apply(null, arguments);
  };
}

var wasmBinaryFile;
  wasmBinaryFile = 'blst.wasm';
  if (!isDataURI(wasmBinaryFile)) {
    wasmBinaryFile = locateFile(wasmBinaryFile);
  }

function getBinary(file) {
  try {
    if (file == wasmBinaryFile && wasmBinary) {
      return new Uint8Array(wasmBinary);
    }
    if (readBinary) {
      return readBinary(file);
    } else {
      throw "both async and sync fetching of the wasm failed";
    }
  }
  catch (err) {
    abort(err);
  }
}

function getBinaryPromise() {
  // If we don't have the binary yet, try to to load it asynchronously.
  // Fetch has some additional restrictions over XHR, like it can't be used on a file:// url.
  // See https://github.com/github/fetch/pull/92#issuecomment-140665932
  // Cordova or Electron apps are typically loaded from a file:// url.
  // So use fetch if it is available and the url is not a file, otherwise fall back to XHR.
  if (!wasmBinary && (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER)) {
    if (typeof fetch == 'function'
      && !isFileURI(wasmBinaryFile)
    ) {
      return fetch(wasmBinaryFile, { credentials: 'same-origin' }).then(function(response) {
        if (!response['ok']) {
          throw "failed to load wasm binary file at '" + wasmBinaryFile + "'";
        }
        return response['arrayBuffer']();
      }).catch(function () {
          return getBinary(wasmBinaryFile);
      });
    }
    else {
      if (readAsync) {
        // fetch is not available or url is file => try XHR (readAsync uses XHR internally)
        return new Promise(function(resolve, reject) {
          readAsync(wasmBinaryFile, function(response) { resolve(new Uint8Array(/** @type{!ArrayBuffer} */(response))) }, reject)
        });
      }
    }
  }

  // Otherwise, getBinary should be able to get it synchronously
  return Promise.resolve().then(function() { return getBinary(wasmBinaryFile); });
}

// Create the wasm instance.
// Receives the wasm imports, returns the exports.
function createWasm() {
  // prepare imports
  var info = {
    'env': asmLibraryArg,
    'wasi_snapshot_preview1': asmLibraryArg,
  };
  // Load the wasm module and create an instance of using native support in the JS engine.
  // handle a generated wasm instance, receiving its exports and
  // performing other necessary setup
  /** @param {WebAssembly.Module=} module*/
  function receiveInstance(instance, module) {
    var exports = instance.exports;

    Module['asm'] = exports;

    wasmMemory = Module['asm']['memory'];
    assert(wasmMemory, "memory not found in wasm exports");
    // This assertion doesn't hold when emscripten is run in --post-link
    // mode.
    // TODO(sbc): Read INITIAL_MEMORY out of the wasm file in post-link mode.
    //assert(wasmMemory.buffer.byteLength === 16777216);
    updateGlobalBufferAndViews(wasmMemory.buffer);

    wasmTable = Module['asm']['__indirect_function_table'];
    assert(wasmTable, "table not found in wasm exports");

    addOnInit(Module['asm']['__wasm_call_ctors']);

    removeRunDependency('wasm-instantiate');

  }
  // we can't run yet (except in a pthread, where we have a custom sync instantiator)
  addRunDependency('wasm-instantiate');

  // Prefer streaming instantiation if available.
  // Async compilation can be confusing when an error on the page overwrites Module
  // (for example, if the order of elements is wrong, and the one defining Module is
  // later), so we save Module and check it later.
  var trueModule = Module;
  function receiveInstantiationResult(result) {
    // 'result' is a ResultObject object which has both the module and instance.
    // receiveInstance() will swap in the exports (to Module.asm) so they can be called
    assert(Module === trueModule, 'the Module object should not be replaced during async compilation - perhaps the order of HTML elements is wrong?');
    trueModule = null;
    // TODO: Due to Closure regression https://github.com/google/closure-compiler/issues/3193, the above line no longer optimizes out down to the following line.
    // When the regression is fixed, can restore the above USE_PTHREADS-enabled path.
    receiveInstance(result['instance']);
  }

  function instantiateArrayBuffer(receiver) {
    return getBinaryPromise().then(function(binary) {
      return WebAssembly.instantiate(binary, info);
    }).then(function (instance) {
      return instance;
    }).then(receiver, function(reason) {
      err('failed to asynchronously prepare wasm: ' + reason);

      // Warn on some common problems.
      if (isFileURI(wasmBinaryFile)) {
        err('warning: Loading from a file URI (' + wasmBinaryFile + ') is not supported in most browsers. See https://emscripten.org/docs/getting_started/FAQ.html#how-do-i-run-a-local-webserver-for-testing-why-does-my-program-stall-in-downloading-or-preparing');
      }
      abort(reason);
    });
  }

  function instantiateAsync() {
    if (!wasmBinary &&
        typeof WebAssembly.instantiateStreaming == 'function' &&
        !isDataURI(wasmBinaryFile) &&
        // Don't use streaming for file:// delivered objects in a webview, fetch them synchronously.
        !isFileURI(wasmBinaryFile) &&
        // Avoid instantiateStreaming() on Node.js environment for now, as while
        // Node.js v18.1.0 implements it, it does not have a full fetch()
        // implementation yet.
        //
        // Reference:
        //   https://github.com/emscripten-core/emscripten/pull/16917
        !ENVIRONMENT_IS_NODE &&
        typeof fetch == 'function') {
      return fetch(wasmBinaryFile, { credentials: 'same-origin' }).then(function(response) {
        // Suppress closure warning here since the upstream definition for
        // instantiateStreaming only allows Promise<Repsponse> rather than
        // an actual Response.
        // TODO(https://github.com/google/closure-compiler/pull/3913): Remove if/when upstream closure is fixed.
        /** @suppress {checkTypes} */
        var result = WebAssembly.instantiateStreaming(response, info);

        return result.then(
          receiveInstantiationResult,
          function(reason) {
            // We expect the most common failure cause to be a bad MIME type for the binary,
            // in which case falling back to ArrayBuffer instantiation should work.
            err('wasm streaming compile failed: ' + reason);
            err('falling back to ArrayBuffer instantiation');
            return instantiateArrayBuffer(receiveInstantiationResult);
          });
      });
    } else {
      return instantiateArrayBuffer(receiveInstantiationResult);
    }
  }

  // User shell pages can write their own Module.instantiateWasm = function(imports, successCallback) callback
  // to manually instantiate the Wasm module themselves. This allows pages to run the instantiation parallel
  // to any other async startup actions they are performing.
  // Also pthreads and wasm workers initialize the wasm instance through this path.
  if (Module['instantiateWasm']) {
    try {
      var exports = Module['instantiateWasm'](info, receiveInstance);
      return exports;
    } catch(e) {
      err('Module.instantiateWasm callback failed with error: ' + e);
      return false;
    }
  }

  instantiateAsync();
  return {}; // no exports yet; we'll fill them in later
}

// Globals used by JS i64 conversions (see makeSetValue)
var tempDouble;
var tempI64;

// === Body ===

var ASM_CONSTS = {
  
};
function array_bounds_check_error(idx,size) { throw 'Array index ' + idx + ' out of bounds: [0,' + size + ')'; }





  function callRuntimeCallbacks(callbacks) {
      while (callbacks.length > 0) {
        var callback = callbacks.shift();
        if (typeof callback == 'function') {
          callback(Module); // Pass the module as the first argument.
          continue;
        }
        var func = callback.func;
        if (typeof func == 'number') {
          if (callback.arg === undefined) {
            // Run the wasm function ptr with signature 'v'. If no function
            // with such signature was exported, this call does not need
            // to be emitted (and would confuse Closure)
            getWasmTableEntry(func)();
          } else {
            // If any function with signature 'vi' was exported, run
            // the callback with that signature.
            getWasmTableEntry(func)(callback.arg);
          }
        } else {
          func(callback.arg === undefined ? null : callback.arg);
        }
      }
    }

  function withStackSave(f) {
      var stack = stackSave();
      var ret = f();
      stackRestore(stack);
      return ret;
    }
  function demangle(func) {
      warnOnce('warning: build with -sDEMANGLE_SUPPORT to link in libcxxabi demangling');
      return func;
    }

  function demangleAll(text) {
      var regex =
        /\b_Z[\w\d_]+/g;
      return text.replace(regex,
        function(x) {
          var y = demangle(x);
          return x === y ? x : (y + ' [' + x + ']');
        });
    }

  
    /**
     * @param {number} ptr
     * @param {string} type
     */
  function getValue(ptr, type = 'i8') {
      if (type.endsWith('*')) type = 'i32';
      switch (type) {
        case 'i1': return HEAP8[((ptr)>>0)];
        case 'i8': return HEAP8[((ptr)>>0)];
        case 'i16': return HEAP16[((ptr)>>1)];
        case 'i32': return HEAP32[((ptr)>>2)];
        case 'i64': return HEAP32[((ptr)>>2)];
        case 'float': return HEAPF32[((ptr)>>2)];
        case 'double': return Number(HEAPF64[((ptr)>>3)]);
        default: abort('invalid type for getValue: ' + type);
      }
      return null;
    }

  var wasmTableMirror = [];
  function getWasmTableEntry(funcPtr) {
      var func = wasmTableMirror[funcPtr];
      if (!func) {
        if (funcPtr >= wasmTableMirror.length) wasmTableMirror.length = funcPtr + 1;
        wasmTableMirror[funcPtr] = func = wasmTable.get(funcPtr);
      }
      assert(wasmTable.get(funcPtr) == func, "JavaScript-side Wasm function table mirror is out of date!");
      return func;
    }

  function handleException(e) {
      // Certain exception types we do not treat as errors since they are used for
      // internal control flow.
      // 1. ExitStatus, which is thrown by exit()
      // 2. "unwind", which is thrown by emscripten_unwind_to_js_event_loop() and others
      //    that wish to return to JS event loop.
      if (e instanceof ExitStatus || e == 'unwind') {
        return EXITSTATUS;
      }
      quit_(1, e);
    }

  function jsStackTrace() {
      var error = new Error();
      if (!error.stack) {
        // IE10+ special cases: It does have callstack info, but it is only
        // populated if an Error object is thrown, so try that as a special-case.
        try {
          throw new Error();
        } catch(e) {
          error = e;
        }
        if (!error.stack) {
          return '(no stack trace available)';
        }
      }
      return error.stack.toString();
    }

  
    /**
     * @param {number} ptr
     * @param {number} value
     * @param {string} type
     */
  function setValue(ptr, value, type = 'i8') {
      if (type.endsWith('*')) type = 'i32';
      switch (type) {
        case 'i1': HEAP8[((ptr)>>0)] = value; break;
        case 'i8': HEAP8[((ptr)>>0)] = value; break;
        case 'i16': HEAP16[((ptr)>>1)] = value; break;
        case 'i32': HEAP32[((ptr)>>2)] = value; break;
        case 'i64': (tempI64 = [value>>>0,(tempDouble=value,(+(Math.abs(tempDouble))) >= 1.0 ? (tempDouble > 0.0 ? ((Math.min((+(Math.floor((tempDouble)/4294967296.0))), 4294967295.0))|0)>>>0 : (~~((+(Math.ceil((tempDouble - +(((~~(tempDouble)))>>>0))/4294967296.0)))))>>>0) : 0)],HEAP32[((ptr)>>2)] = tempI64[0],HEAP32[(((ptr)+(4))>>2)] = tempI64[1]); break;
        case 'float': HEAPF32[((ptr)>>2)] = value; break;
        case 'double': HEAPF64[((ptr)>>3)] = value; break;
        default: abort('invalid type for setValue: ' + type);
      }
    }

  function setWasmTableEntry(idx, func) {
      wasmTable.set(idx, func);
      // With ABORT_ON_WASM_EXCEPTIONS wasmTable.get is overriden to return wrapped
      // functions so we need to call it here to retrieve the potential wrapper correctly
      // instead of just storing 'func' directly into wasmTableMirror
      wasmTableMirror[idx] = wasmTable.get(idx);
    }

  function stackTrace() {
      var js = jsStackTrace();
      if (Module['extraStackTrace']) js += '\n' + Module['extraStackTrace']();
      return demangleAll(js);
    }

  function ___cxa_allocate_exception(size) {
      // Thrown object is prepended by exception metadata block
      return _malloc(size + 24) + 24;
    }

  /** @constructor */
  function ExceptionInfo(excPtr) {
      this.excPtr = excPtr;
      this.ptr = excPtr - 24;
  
      this.set_type = function(type) {
        HEAPU32[(((this.ptr)+(4))>>2)] = type;
      };
  
      this.get_type = function() {
        return HEAPU32[(((this.ptr)+(4))>>2)];
      };
  
      this.set_destructor = function(destructor) {
        HEAPU32[(((this.ptr)+(8))>>2)] = destructor;
      };
  
      this.get_destructor = function() {
        return HEAPU32[(((this.ptr)+(8))>>2)];
      };
  
      this.set_refcount = function(refcount) {
        HEAP32[((this.ptr)>>2)] = refcount;
      };
  
      this.set_caught = function (caught) {
        caught = caught ? 1 : 0;
        HEAP8[(((this.ptr)+(12))>>0)] = caught;
      };
  
      this.get_caught = function () {
        return HEAP8[(((this.ptr)+(12))>>0)] != 0;
      };
  
      this.set_rethrown = function (rethrown) {
        rethrown = rethrown ? 1 : 0;
        HEAP8[(((this.ptr)+(13))>>0)] = rethrown;
      };
  
      this.get_rethrown = function () {
        return HEAP8[(((this.ptr)+(13))>>0)] != 0;
      };
  
      // Initialize native structure fields. Should be called once after allocated.
      this.init = function(type, destructor) {
        this.set_adjusted_ptr(0);
        this.set_type(type);
        this.set_destructor(destructor);
        this.set_refcount(0);
        this.set_caught(false);
        this.set_rethrown(false);
      }
  
      this.add_ref = function() {
        var value = HEAP32[((this.ptr)>>2)];
        HEAP32[((this.ptr)>>2)] = value + 1;
      };
  
      // Returns true if last reference released.
      this.release_ref = function() {
        var prev = HEAP32[((this.ptr)>>2)];
        HEAP32[((this.ptr)>>2)] = prev - 1;
        assert(prev > 0);
        return prev === 1;
      };
  
      this.set_adjusted_ptr = function(adjustedPtr) {
        HEAPU32[(((this.ptr)+(16))>>2)] = adjustedPtr;
      };
  
      this.get_adjusted_ptr = function() {
        return HEAPU32[(((this.ptr)+(16))>>2)];
      };
  
      // Get pointer which is expected to be received by catch clause in C++ code. It may be adjusted
      // when the pointer is casted to some of the exception object base classes (e.g. when virtual
      // inheritance is used). When a pointer is thrown this method should return the thrown pointer
      // itself.
      this.get_exception_ptr = function() {
        // Work around a fastcomp bug, this code is still included for some reason in a build without
        // exceptions support.
        var isPointer = ___cxa_is_pointer_type(this.get_type());
        if (isPointer) {
          return HEAPU32[((this.excPtr)>>2)];
        }
        var adjusted = this.get_adjusted_ptr();
        if (adjusted !== 0) return adjusted;
        return this.excPtr;
      };
    }
  
  var exceptionLast = 0;
  
  var uncaughtExceptionCount = 0;
  function ___cxa_throw(ptr, type, destructor) {
      var info = new ExceptionInfo(ptr);
      // Initialize ExceptionInfo content after it was allocated in __cxa_allocate_exception.
      info.init(type, destructor);
      exceptionLast = ptr;
      uncaughtExceptionCount++;
      throw ptr + " - Exception catching is disabled, this exception cannot be caught. Compile with -sNO_DISABLE_EXCEPTION_CATCHING or -sEXCEPTION_CATCHING_ALLOWED=[..] to catch.";
    }

  function _abort() {
      abort('native code called abort()');
    }

  function _emscripten_memcpy_big(dest, src, num) {
      HEAPU8.copyWithin(dest, src, src + num);
    }

  function getHeapMax() {
      return HEAPU8.length;
    }
  
  function abortOnCannotGrowMemory(requestedSize) {
      abort('Cannot enlarge memory arrays to size ' + requestedSize + ' bytes (OOM). Either (1) compile with -sINITIAL_MEMORY=X with X higher than the current value ' + HEAP8.length + ', (2) compile with -sALLOW_MEMORY_GROWTH which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with -sABORTING_MALLOC=0');
    }
  function _emscripten_resize_heap(requestedSize) {
      var oldSize = HEAPU8.length;
      requestedSize = requestedSize >>> 0;
      abortOnCannotGrowMemory(requestedSize);
    }

  var SYSCALLS = {varargs:undefined,get:function() {
        assert(SYSCALLS.varargs != undefined);
        SYSCALLS.varargs += 4;
        var ret = HEAP32[(((SYSCALLS.varargs)-(4))>>2)];
        return ret;
      },getStr:function(ptr) {
        var ret = UTF8ToString(ptr);
        return ret;
      }};
  function _fd_close(fd) {
      abort('fd_close called without SYSCALLS_REQUIRE_FILESYSTEM');
    }

  function convertI32PairToI53Checked(lo, hi) {
      assert(lo == (lo >>> 0) || lo == (lo|0)); // lo should either be a i32 or a u32
      assert(hi === (hi|0));                    // hi should be a i32
      return ((hi + 0x200000) >>> 0 < 0x400001 - !!lo) ? (lo >>> 0) + hi * 4294967296 : NaN;
    }
  function _fd_seek(fd, offset_low, offset_high, whence, newOffset) {
      return 70;
    }

  var printCharBuffers = [null,[],[]];
  function printChar(stream, curr) {
      var buffer = printCharBuffers[stream];
      assert(buffer);
      if (curr === 0 || curr === 10) {
        (stream === 1 ? out : err)(UTF8ArrayToString(buffer, 0));
        buffer.length = 0;
      } else {
        buffer.push(curr);
      }
    }
  function flush_NO_FILESYSTEM() {
      // flush anything remaining in the buffers during shutdown
      _fflush(0);
      if (printCharBuffers[1].length) printChar(1, 10);
      if (printCharBuffers[2].length) printChar(2, 10);
    }
  function _fd_write(fd, iov, iovcnt, pnum) {
      // hack to support printf in SYSCALLS_REQUIRE_FILESYSTEM=0
      var num = 0;
      for (var i = 0; i < iovcnt; i++) {
        var ptr = HEAPU32[((iov)>>2)];
        var len = HEAPU32[(((iov)+(4))>>2)];
        iov += 8;
        for (var j = 0; j < len; j++) {
          printChar(fd, HEAPU8[ptr+j]);
        }
        num += len;
      }
      HEAPU32[((pnum)>>2)] = num;
      return 0;
    }

  function _setTempRet0(val) {
      setTempRet0(val);
    }
var ASSERTIONS = true;



/** @type {function(string, boolean=, number=)} */
function intArrayFromString(stringy, dontAddNull, length) {
  var len = length > 0 ? length : lengthBytesUTF8(stringy)+1;
  var u8array = new Array(len);
  var numBytesWritten = stringToUTF8Array(stringy, u8array, 0, u8array.length);
  if (dontAddNull) u8array.length = numBytesWritten;
  return u8array;
}

function intArrayToString(array) {
  var ret = [];
  for (var i = 0; i < array.length; i++) {
    var chr = array[i];
    if (chr > 0xFF) {
      if (ASSERTIONS) {
        assert(false, 'Character code ' + chr + ' (' + String.fromCharCode(chr) + ')  at offset ' + i + ' not in 0x00-0xFF.');
      }
      chr &= 0xFF;
    }
    ret.push(String.fromCharCode(chr));
  }
  return ret.join('');
}


function checkIncomingModuleAPI() {
  ignoredModuleProp('fetchSettings');
}
var asmLibraryArg = {
  "__cxa_allocate_exception": ___cxa_allocate_exception,
  "__cxa_throw": ___cxa_throw,
  "abort": _abort,
  "array_bounds_check_error": array_bounds_check_error,
  "emscripten_memcpy_big": _emscripten_memcpy_big,
  "emscripten_resize_heap": _emscripten_resize_heap,
  "fd_close": _fd_close,
  "fd_seek": _fd_seek,
  "fd_write": _fd_write,
  "setTempRet0": _setTempRet0
};
var asm = createWasm();
/** @type {function(...*):?} */
var ___wasm_call_ctors = Module["___wasm_call_ctors"] = createExportWrapper("__wasm_call_ctors");

/** @type {function(...*):?} */
var _emscripten_bind_VoidPtr___destroy___0 = Module["_emscripten_bind_VoidPtr___destroy___0"] = createExportWrapper("emscripten_bind_VoidPtr___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_P1_Affine_0 = Module["_emscripten_bind_P1_Affine_P1_Affine_0"] = createExportWrapper("emscripten_bind_P1_Affine_P1_Affine_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_P1_Affine_1 = Module["_emscripten_bind_P1_Affine_P1_Affine_1"] = createExportWrapper("emscripten_bind_P1_Affine_P1_Affine_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_P1_Affine_2 = Module["_emscripten_bind_P1_Affine_P1_Affine_2"] = createExportWrapper("emscripten_bind_P1_Affine_P1_Affine_2");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_dup_0 = Module["_emscripten_bind_P1_Affine_dup_0"] = createExportWrapper("emscripten_bind_P1_Affine_dup_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_to_jacobian_0 = Module["_emscripten_bind_P1_Affine_to_jacobian_0"] = createExportWrapper("emscripten_bind_P1_Affine_to_jacobian_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_serialize_1 = Module["_emscripten_bind_P1_Affine_serialize_1"] = createExportWrapper("emscripten_bind_P1_Affine_serialize_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_compress_1 = Module["_emscripten_bind_P1_Affine_compress_1"] = createExportWrapper("emscripten_bind_P1_Affine_compress_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_on_curve_0 = Module["_emscripten_bind_P1_Affine_on_curve_0"] = createExportWrapper("emscripten_bind_P1_Affine_on_curve_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_in_group_0 = Module["_emscripten_bind_P1_Affine_in_group_0"] = createExportWrapper("emscripten_bind_P1_Affine_in_group_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_is_inf_0 = Module["_emscripten_bind_P1_Affine_is_inf_0"] = createExportWrapper("emscripten_bind_P1_Affine_is_inf_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_is_equal_1 = Module["_emscripten_bind_P1_Affine_is_equal_1"] = createExportWrapper("emscripten_bind_P1_Affine_is_equal_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine_core_verify_7 = Module["_emscripten_bind_P1_Affine_core_verify_7"] = createExportWrapper("emscripten_bind_P1_Affine_core_verify_7");

/** @type {function(...*):?} */
var _emscripten_bind_P1_Affine___destroy___0 = Module["_emscripten_bind_P1_Affine___destroy___0"] = createExportWrapper("emscripten_bind_P1_Affine___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_P1_0 = Module["_emscripten_bind_P1_P1_0"] = createExportWrapper("emscripten_bind_P1_P1_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_P1_1 = Module["_emscripten_bind_P1_P1_1"] = createExportWrapper("emscripten_bind_P1_P1_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_P1_2 = Module["_emscripten_bind_P1_P1_2"] = createExportWrapper("emscripten_bind_P1_P1_2");

/** @type {function(...*):?} */
var _emscripten_bind_P1_P1_generator_0 = Module["_emscripten_bind_P1_P1_generator_0"] = createExportWrapper("emscripten_bind_P1_P1_generator_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_dup_0 = Module["_emscripten_bind_P1_dup_0"] = createExportWrapper("emscripten_bind_P1_dup_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_to_affine_0 = Module["_emscripten_bind_P1_to_affine_0"] = createExportWrapper("emscripten_bind_P1_to_affine_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_serialize_1 = Module["_emscripten_bind_P1_serialize_1"] = createExportWrapper("emscripten_bind_P1_serialize_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_compress_1 = Module["_emscripten_bind_P1_compress_1"] = createExportWrapper("emscripten_bind_P1_compress_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_on_curve_0 = Module["_emscripten_bind_P1_on_curve_0"] = createExportWrapper("emscripten_bind_P1_on_curve_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_in_group_0 = Module["_emscripten_bind_P1_in_group_0"] = createExportWrapper("emscripten_bind_P1_in_group_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_is_inf_0 = Module["_emscripten_bind_P1_is_inf_0"] = createExportWrapper("emscripten_bind_P1_is_inf_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_is_equal_1 = Module["_emscripten_bind_P1_is_equal_1"] = createExportWrapper("emscripten_bind_P1_is_equal_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_aggregate_1 = Module["_emscripten_bind_P1_aggregate_1"] = createExportWrapper("emscripten_bind_P1_aggregate_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_sign_with_1 = Module["_emscripten_bind_P1_sign_with_1"] = createExportWrapper("emscripten_bind_P1_sign_with_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_hash_to_5 = Module["_emscripten_bind_P1_hash_to_5"] = createExportWrapper("emscripten_bind_P1_hash_to_5");

/** @type {function(...*):?} */
var _emscripten_bind_P1_encode_to_5 = Module["_emscripten_bind_P1_encode_to_5"] = createExportWrapper("emscripten_bind_P1_encode_to_5");

/** @type {function(...*):?} */
var _emscripten_bind_P1_mult_1 = Module["_emscripten_bind_P1_mult_1"] = createExportWrapper("emscripten_bind_P1_mult_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_neg_0 = Module["_emscripten_bind_P1_neg_0"] = createExportWrapper("emscripten_bind_P1_neg_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1_cneg_1 = Module["_emscripten_bind_P1_cneg_1"] = createExportWrapper("emscripten_bind_P1_cneg_1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_add_1__P1 = Module["_emscripten_bind_P1_add_1__P1"] = createExportWrapper("emscripten_bind_P1_add_1__P1");

/** @type {function(...*):?} */
var _emscripten_bind_P1_add_1__P1_Affine = Module["_emscripten_bind_P1_add_1__P1_Affine"] = createExportWrapper("emscripten_bind_P1_add_1__P1_Affine");

/** @type {function(...*):?} */
var _emscripten_bind_P1_dbl_0 = Module["_emscripten_bind_P1_dbl_0"] = createExportWrapper("emscripten_bind_P1_dbl_0");

/** @type {function(...*):?} */
var _emscripten_bind_P1___destroy___0 = Module["_emscripten_bind_P1___destroy___0"] = createExportWrapper("emscripten_bind_P1___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_P2_Affine_0 = Module["_emscripten_bind_P2_Affine_P2_Affine_0"] = createExportWrapper("emscripten_bind_P2_Affine_P2_Affine_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_P2_Affine_1 = Module["_emscripten_bind_P2_Affine_P2_Affine_1"] = createExportWrapper("emscripten_bind_P2_Affine_P2_Affine_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_P2_Affine_2 = Module["_emscripten_bind_P2_Affine_P2_Affine_2"] = createExportWrapper("emscripten_bind_P2_Affine_P2_Affine_2");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_dup_0 = Module["_emscripten_bind_P2_Affine_dup_0"] = createExportWrapper("emscripten_bind_P2_Affine_dup_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_to_jacobian_0 = Module["_emscripten_bind_P2_Affine_to_jacobian_0"] = createExportWrapper("emscripten_bind_P2_Affine_to_jacobian_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_serialize_1 = Module["_emscripten_bind_P2_Affine_serialize_1"] = createExportWrapper("emscripten_bind_P2_Affine_serialize_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_compress_1 = Module["_emscripten_bind_P2_Affine_compress_1"] = createExportWrapper("emscripten_bind_P2_Affine_compress_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_on_curve_0 = Module["_emscripten_bind_P2_Affine_on_curve_0"] = createExportWrapper("emscripten_bind_P2_Affine_on_curve_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_in_group_0 = Module["_emscripten_bind_P2_Affine_in_group_0"] = createExportWrapper("emscripten_bind_P2_Affine_in_group_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_is_inf_0 = Module["_emscripten_bind_P2_Affine_is_inf_0"] = createExportWrapper("emscripten_bind_P2_Affine_is_inf_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_is_equal_1 = Module["_emscripten_bind_P2_Affine_is_equal_1"] = createExportWrapper("emscripten_bind_P2_Affine_is_equal_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine_core_verify_7 = Module["_emscripten_bind_P2_Affine_core_verify_7"] = createExportWrapper("emscripten_bind_P2_Affine_core_verify_7");

/** @type {function(...*):?} */
var _emscripten_bind_P2_Affine___destroy___0 = Module["_emscripten_bind_P2_Affine___destroy___0"] = createExportWrapper("emscripten_bind_P2_Affine___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_P2_0 = Module["_emscripten_bind_P2_P2_0"] = createExportWrapper("emscripten_bind_P2_P2_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_P2_1 = Module["_emscripten_bind_P2_P2_1"] = createExportWrapper("emscripten_bind_P2_P2_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_P2_2 = Module["_emscripten_bind_P2_P2_2"] = createExportWrapper("emscripten_bind_P2_P2_2");

/** @type {function(...*):?} */
var _emscripten_bind_P2_P2_generator_0 = Module["_emscripten_bind_P2_P2_generator_0"] = createExportWrapper("emscripten_bind_P2_P2_generator_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_dup_0 = Module["_emscripten_bind_P2_dup_0"] = createExportWrapper("emscripten_bind_P2_dup_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_to_affine_0 = Module["_emscripten_bind_P2_to_affine_0"] = createExportWrapper("emscripten_bind_P2_to_affine_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_serialize_1 = Module["_emscripten_bind_P2_serialize_1"] = createExportWrapper("emscripten_bind_P2_serialize_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_compress_1 = Module["_emscripten_bind_P2_compress_1"] = createExportWrapper("emscripten_bind_P2_compress_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_on_curve_0 = Module["_emscripten_bind_P2_on_curve_0"] = createExportWrapper("emscripten_bind_P2_on_curve_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_in_group_0 = Module["_emscripten_bind_P2_in_group_0"] = createExportWrapper("emscripten_bind_P2_in_group_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_is_inf_0 = Module["_emscripten_bind_P2_is_inf_0"] = createExportWrapper("emscripten_bind_P2_is_inf_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_is_equal_1 = Module["_emscripten_bind_P2_is_equal_1"] = createExportWrapper("emscripten_bind_P2_is_equal_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_aggregate_1 = Module["_emscripten_bind_P2_aggregate_1"] = createExportWrapper("emscripten_bind_P2_aggregate_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_sign_with_1 = Module["_emscripten_bind_P2_sign_with_1"] = createExportWrapper("emscripten_bind_P2_sign_with_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_hash_to_5 = Module["_emscripten_bind_P2_hash_to_5"] = createExportWrapper("emscripten_bind_P2_hash_to_5");

/** @type {function(...*):?} */
var _emscripten_bind_P2_encode_to_5 = Module["_emscripten_bind_P2_encode_to_5"] = createExportWrapper("emscripten_bind_P2_encode_to_5");

/** @type {function(...*):?} */
var _emscripten_bind_P2_mult_1 = Module["_emscripten_bind_P2_mult_1"] = createExportWrapper("emscripten_bind_P2_mult_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_neg_0 = Module["_emscripten_bind_P2_neg_0"] = createExportWrapper("emscripten_bind_P2_neg_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2_cneg_1 = Module["_emscripten_bind_P2_cneg_1"] = createExportWrapper("emscripten_bind_P2_cneg_1");

/** @type {function(...*):?} */
var _emscripten_bind_P2_add_1__P2 = Module["_emscripten_bind_P2_add_1__P2"] = createExportWrapper("emscripten_bind_P2_add_1__P2");

/** @type {function(...*):?} */
var _emscripten_bind_P2_add_1__P2_Affine = Module["_emscripten_bind_P2_add_1__P2_Affine"] = createExportWrapper("emscripten_bind_P2_add_1__P2_Affine");

/** @type {function(...*):?} */
var _emscripten_bind_P2_dbl_0 = Module["_emscripten_bind_P2_dbl_0"] = createExportWrapper("emscripten_bind_P2_dbl_0");

/** @type {function(...*):?} */
var _emscripten_bind_P2___destroy___0 = Module["_emscripten_bind_P2___destroy___0"] = createExportWrapper("emscripten_bind_P2___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey_SecretKey_0 = Module["_emscripten_bind_SecretKey_SecretKey_0"] = createExportWrapper("emscripten_bind_SecretKey_SecretKey_0");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey_keygen_3 = Module["_emscripten_bind_SecretKey_keygen_3"] = createExportWrapper("emscripten_bind_SecretKey_keygen_3");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey_from_bendian_1 = Module["_emscripten_bind_SecretKey_from_bendian_1"] = createExportWrapper("emscripten_bind_SecretKey_from_bendian_1");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey_from_lendian_1 = Module["_emscripten_bind_SecretKey_from_lendian_1"] = createExportWrapper("emscripten_bind_SecretKey_from_lendian_1");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey_to_bendian_1 = Module["_emscripten_bind_SecretKey_to_bendian_1"] = createExportWrapper("emscripten_bind_SecretKey_to_bendian_1");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey_to_lendian_1 = Module["_emscripten_bind_SecretKey_to_lendian_1"] = createExportWrapper("emscripten_bind_SecretKey_to_lendian_1");

/** @type {function(...*):?} */
var _emscripten_bind_SecretKey___destroy___0 = Module["_emscripten_bind_SecretKey___destroy___0"] = createExportWrapper("emscripten_bind_SecretKey___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_PT_PT__P1_Affine_1 = Module["_emscripten_bind_PT_PT__P1_Affine_1"] = createExportWrapper("emscripten_bind_PT_PT__P1_Affine_1");

/** @type {function(...*):?} */
var _emscripten_bind_PT_PT__P2_Affine_1 = Module["_emscripten_bind_PT_PT__P2_Affine_1"] = createExportWrapper("emscripten_bind_PT_PT__P2_Affine_1");

/** @type {function(...*):?} */
var _emscripten_bind_PT_PT_2 = Module["_emscripten_bind_PT_PT_2"] = createExportWrapper("emscripten_bind_PT_PT_2");

/** @type {function(...*):?} */
var _emscripten_bind_PT_dup_0 = Module["_emscripten_bind_PT_dup_0"] = createExportWrapper("emscripten_bind_PT_dup_0");

/** @type {function(...*):?} */
var _emscripten_bind_PT_is_one_0 = Module["_emscripten_bind_PT_is_one_0"] = createExportWrapper("emscripten_bind_PT_is_one_0");

/** @type {function(...*):?} */
var _emscripten_bind_PT_is_equal_1 = Module["_emscripten_bind_PT_is_equal_1"] = createExportWrapper("emscripten_bind_PT_is_equal_1");

/** @type {function(...*):?} */
var _emscripten_bind_PT_sqr_0 = Module["_emscripten_bind_PT_sqr_0"] = createExportWrapper("emscripten_bind_PT_sqr_0");

/** @type {function(...*):?} */
var _emscripten_bind_PT_mul_1 = Module["_emscripten_bind_PT_mul_1"] = createExportWrapper("emscripten_bind_PT_mul_1");

/** @type {function(...*):?} */
var _emscripten_bind_PT_final_exp_0 = Module["_emscripten_bind_PT_final_exp_0"] = createExportWrapper("emscripten_bind_PT_final_exp_0");

/** @type {function(...*):?} */
var _emscripten_bind_PT_in_group_0 = Module["_emscripten_bind_PT_in_group_0"] = createExportWrapper("emscripten_bind_PT_in_group_0");

/** @type {function(...*):?} */
var _emscripten_bind_PT_finalverify_2 = Module["_emscripten_bind_PT_finalverify_2"] = createExportWrapper("emscripten_bind_PT_finalverify_2");

/** @type {function(...*):?} */
var _emscripten_bind_PT_one_0 = Module["_emscripten_bind_PT_one_0"] = createExportWrapper("emscripten_bind_PT_one_0");

/** @type {function(...*):?} */
var _emscripten_bind_PT___destroy___0 = Module["_emscripten_bind_PT___destroy___0"] = createExportWrapper("emscripten_bind_PT___destroy___0");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_Pairing_3 = Module["_emscripten_bind_Pairing_Pairing_3"] = createExportWrapper("emscripten_bind_Pairing_Pairing_3");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_aggregate_6__P1_P2 = Module["_emscripten_bind_Pairing_aggregate_6__P1_P2"] = createExportWrapper("emscripten_bind_Pairing_aggregate_6__P1_P2");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_aggregate_6__P2_P1 = Module["_emscripten_bind_Pairing_aggregate_6__P2_P1"] = createExportWrapper("emscripten_bind_Pairing_aggregate_6__P2_P1");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_mul_n_aggregate_8__P1_P2 = Module["_emscripten_bind_Pairing_mul_n_aggregate_8__P1_P2"] = createExportWrapper("emscripten_bind_Pairing_mul_n_aggregate_8__P1_P2");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_mul_n_aggregate_8__P2_P1 = Module["_emscripten_bind_Pairing_mul_n_aggregate_8__P2_P1"] = createExportWrapper("emscripten_bind_Pairing_mul_n_aggregate_8__P2_P1");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_commit_0 = Module["_emscripten_bind_Pairing_commit_0"] = createExportWrapper("emscripten_bind_Pairing_commit_0");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_merge_1 = Module["_emscripten_bind_Pairing_merge_1"] = createExportWrapper("emscripten_bind_Pairing_merge_1");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing_finalverify_1 = Module["_emscripten_bind_Pairing_finalverify_1"] = createExportWrapper("emscripten_bind_Pairing_finalverify_1");

/** @type {function(...*):?} */
var _emscripten_bind_Pairing___destroy___0 = Module["_emscripten_bind_Pairing___destroy___0"] = createExportWrapper("emscripten_bind_Pairing___destroy___0");

/** @type {function(...*):?} */
var __Z20emscripten_bind_G1_0v = Module["__Z20emscripten_bind_G1_0v"] = createExportWrapper("_Z20emscripten_bind_G1_0v");

/** @type {function(...*):?} */
var __Z20emscripten_bind_G2_0v = Module["__Z20emscripten_bind_G2_0v"] = createExportWrapper("_Z20emscripten_bind_G2_0v");

/** @type {function(...*):?} */
var ___errno_location = Module["___errno_location"] = createExportWrapper("__errno_location");

/** @type {function(...*):?} */
var _fflush = Module["_fflush"] = createExportWrapper("fflush");

/** @type {function(...*):?} */
var _malloc = Module["_malloc"] = createExportWrapper("malloc");

/** @type {function(...*):?} */
var _emscripten_stack_init = Module["_emscripten_stack_init"] = function() {
  return (_emscripten_stack_init = Module["_emscripten_stack_init"] = Module["asm"]["emscripten_stack_init"]).apply(null, arguments);
};

/** @type {function(...*):?} */
var _emscripten_stack_get_free = Module["_emscripten_stack_get_free"] = function() {
  return (_emscripten_stack_get_free = Module["_emscripten_stack_get_free"] = Module["asm"]["emscripten_stack_get_free"]).apply(null, arguments);
};

/** @type {function(...*):?} */
var _emscripten_stack_get_base = Module["_emscripten_stack_get_base"] = function() {
  return (_emscripten_stack_get_base = Module["_emscripten_stack_get_base"] = Module["asm"]["emscripten_stack_get_base"]).apply(null, arguments);
};

/** @type {function(...*):?} */
var _emscripten_stack_get_end = Module["_emscripten_stack_get_end"] = function() {
  return (_emscripten_stack_get_end = Module["_emscripten_stack_get_end"] = Module["asm"]["emscripten_stack_get_end"]).apply(null, arguments);
};

/** @type {function(...*):?} */
var stackSave = Module["stackSave"] = createExportWrapper("stackSave");

/** @type {function(...*):?} */
var stackRestore = Module["stackRestore"] = createExportWrapper("stackRestore");

/** @type {function(...*):?} */
var stackAlloc = Module["stackAlloc"] = createExportWrapper("stackAlloc");

/** @type {function(...*):?} */
var ___cxa_is_pointer_type = Module["___cxa_is_pointer_type"] = createExportWrapper("__cxa_is_pointer_type");

/** @type {function(...*):?} */
var dynCall_jiji = Module["dynCall_jiji"] = createExportWrapper("dynCall_jiji");





// === Auto-generated postamble setup entry stuff ===

unexportedRuntimeFunction('ccall', false);
unexportedRuntimeFunction('cwrap', false);
unexportedRuntimeFunction('allocate', false);
unexportedRuntimeFunction('UTF8ArrayToString', false);
unexportedRuntimeFunction('UTF8ToString', false);
unexportedRuntimeFunction('stringToUTF8Array', false);
unexportedRuntimeFunction('stringToUTF8', false);
unexportedRuntimeFunction('lengthBytesUTF8', false);
unexportedRuntimeFunction('addOnPreRun', false);
unexportedRuntimeFunction('addOnInit', false);
unexportedRuntimeFunction('addOnPreMain', false);
unexportedRuntimeFunction('addOnExit', false);
unexportedRuntimeFunction('addOnPostRun', false);
unexportedRuntimeFunction('addRunDependency', true);
unexportedRuntimeFunction('removeRunDependency', true);
unexportedRuntimeFunction('FS_createFolder', false);
unexportedRuntimeFunction('FS_createPath', true);
unexportedRuntimeFunction('FS_createDataFile', true);
unexportedRuntimeFunction('FS_createPreloadedFile', true);
unexportedRuntimeFunction('FS_createLazyFile', true);
unexportedRuntimeFunction('FS_createLink', false);
unexportedRuntimeFunction('FS_createDevice', true);
unexportedRuntimeFunction('FS_unlink', true);
unexportedRuntimeFunction('getLEB', false);
unexportedRuntimeFunction('getFunctionTables', false);
unexportedRuntimeFunction('alignFunctionTables', false);
unexportedRuntimeFunction('registerFunctions', false);
unexportedRuntimeFunction('addFunction', false);
unexportedRuntimeFunction('removeFunction', false);
unexportedRuntimeFunction('prettyPrint', false);
unexportedRuntimeFunction('getCompilerSetting', false);
unexportedRuntimeFunction('print', false);
unexportedRuntimeFunction('printErr', false);
unexportedRuntimeFunction('getTempRet0', false);
unexportedRuntimeFunction('setTempRet0', false);
unexportedRuntimeFunction('callMain', false);
unexportedRuntimeFunction('abort', false);
unexportedRuntimeFunction('keepRuntimeAlive', false);
unexportedRuntimeFunction('wasmMemory', false);
unexportedRuntimeFunction('warnOnce', false);
unexportedRuntimeFunction('stackSave', false);
unexportedRuntimeFunction('stackRestore', false);
unexportedRuntimeFunction('stackAlloc', false);
unexportedRuntimeFunction('AsciiToString', false);
unexportedRuntimeFunction('stringToAscii', false);
unexportedRuntimeFunction('UTF16ToString', false);
unexportedRuntimeFunction('stringToUTF16', false);
unexportedRuntimeFunction('lengthBytesUTF16', false);
unexportedRuntimeFunction('UTF32ToString', false);
unexportedRuntimeFunction('stringToUTF32', false);
unexportedRuntimeFunction('lengthBytesUTF32', false);
unexportedRuntimeFunction('allocateUTF8', false);
unexportedRuntimeFunction('allocateUTF8OnStack', false);
unexportedRuntimeFunction('ExitStatus', false);
unexportedRuntimeFunction('intArrayFromString', false);
unexportedRuntimeFunction('intArrayToString', false);
unexportedRuntimeFunction('writeStringToMemory', false);
unexportedRuntimeFunction('writeArrayToMemory', false);
unexportedRuntimeFunction('writeAsciiToMemory', false);
Module["writeStackCookie"] = writeStackCookie;
Module["checkStackCookie"] = checkStackCookie;
unexportedRuntimeFunction('ptrToString', false);
unexportedRuntimeFunction('zeroMemory', false);
unexportedRuntimeFunction('stringToNewUTF8', false);
unexportedRuntimeFunction('getHeapMax', false);
unexportedRuntimeFunction('abortOnCannotGrowMemory', false);
unexportedRuntimeFunction('emscripten_realloc_buffer', false);
unexportedRuntimeFunction('ENV', false);
unexportedRuntimeFunction('ERRNO_CODES', false);
unexportedRuntimeFunction('ERRNO_MESSAGES', false);
unexportedRuntimeFunction('setErrNo', false);
unexportedRuntimeFunction('inetPton4', false);
unexportedRuntimeFunction('inetNtop4', false);
unexportedRuntimeFunction('inetPton6', false);
unexportedRuntimeFunction('inetNtop6', false);
unexportedRuntimeFunction('readSockaddr', false);
unexportedRuntimeFunction('writeSockaddr', false);
unexportedRuntimeFunction('DNS', false);
unexportedRuntimeFunction('getHostByName', false);
unexportedRuntimeFunction('Protocols', false);
unexportedRuntimeFunction('Sockets', false);
unexportedRuntimeFunction('getRandomDevice', false);
unexportedRuntimeFunction('traverseStack', false);
unexportedRuntimeFunction('UNWIND_CACHE', false);
unexportedRuntimeFunction('convertPCtoSourceLocation', false);
unexportedRuntimeFunction('readAsmConstArgsArray', false);
unexportedRuntimeFunction('readAsmConstArgs', false);
unexportedRuntimeFunction('mainThreadEM_ASM', false);
unexportedRuntimeFunction('jstoi_q', false);
unexportedRuntimeFunction('jstoi_s', false);
unexportedRuntimeFunction('getExecutableName', false);
unexportedRuntimeFunction('listenOnce', false);
unexportedRuntimeFunction('autoResumeAudioContext', false);
unexportedRuntimeFunction('dynCallLegacy', false);
unexportedRuntimeFunction('getDynCaller', false);
unexportedRuntimeFunction('dynCall', false);
unexportedRuntimeFunction('handleException', false);
unexportedRuntimeFunction('runtimeKeepalivePush', false);
unexportedRuntimeFunction('runtimeKeepalivePop', false);
unexportedRuntimeFunction('callUserCallback', false);
unexportedRuntimeFunction('maybeExit', false);
unexportedRuntimeFunction('safeSetTimeout', false);
unexportedRuntimeFunction('asmjsMangle', false);
unexportedRuntimeFunction('asyncLoad', false);
unexportedRuntimeFunction('alignMemory', false);
unexportedRuntimeFunction('mmapAlloc', false);
unexportedRuntimeFunction('writeI53ToI64', false);
unexportedRuntimeFunction('writeI53ToI64Clamped', false);
unexportedRuntimeFunction('writeI53ToI64Signaling', false);
unexportedRuntimeFunction('writeI53ToU64Clamped', false);
unexportedRuntimeFunction('writeI53ToU64Signaling', false);
unexportedRuntimeFunction('readI53FromI64', false);
unexportedRuntimeFunction('readI53FromU64', false);
unexportedRuntimeFunction('convertI32PairToI53', false);
unexportedRuntimeFunction('convertI32PairToI53Checked', false);
unexportedRuntimeFunction('convertU32PairToI53', false);
unexportedRuntimeFunction('reallyNegative', false);
unexportedRuntimeFunction('unSign', false);
unexportedRuntimeFunction('strLen', false);
unexportedRuntimeFunction('reSign', false);
unexportedRuntimeFunction('formatString', false);
unexportedRuntimeFunction('setValue', false);
unexportedRuntimeFunction('getValue', false);
unexportedRuntimeFunction('PATH', false);
unexportedRuntimeFunction('PATH_FS', false);
unexportedRuntimeFunction('SYSCALLS', false);
unexportedRuntimeFunction('getSocketFromFD', false);
unexportedRuntimeFunction('getSocketAddress', false);
unexportedRuntimeFunction('JSEvents', false);
unexportedRuntimeFunction('registerKeyEventCallback', false);
unexportedRuntimeFunction('specialHTMLTargets', false);
unexportedRuntimeFunction('maybeCStringToJsString', false);
unexportedRuntimeFunction('findEventTarget', false);
unexportedRuntimeFunction('findCanvasEventTarget', false);
unexportedRuntimeFunction('getBoundingClientRect', false);
unexportedRuntimeFunction('fillMouseEventData', false);
unexportedRuntimeFunction('registerMouseEventCallback', false);
unexportedRuntimeFunction('registerWheelEventCallback', false);
unexportedRuntimeFunction('registerUiEventCallback', false);
unexportedRuntimeFunction('registerFocusEventCallback', false);
unexportedRuntimeFunction('fillDeviceOrientationEventData', false);
unexportedRuntimeFunction('registerDeviceOrientationEventCallback', false);
unexportedRuntimeFunction('fillDeviceMotionEventData', false);
unexportedRuntimeFunction('registerDeviceMotionEventCallback', false);
unexportedRuntimeFunction('screenOrientation', false);
unexportedRuntimeFunction('fillOrientationChangeEventData', false);
unexportedRuntimeFunction('registerOrientationChangeEventCallback', false);
unexportedRuntimeFunction('fillFullscreenChangeEventData', false);
unexportedRuntimeFunction('registerFullscreenChangeEventCallback', false);
unexportedRuntimeFunction('JSEvents_requestFullscreen', false);
unexportedRuntimeFunction('JSEvents_resizeCanvasForFullscreen', false);
unexportedRuntimeFunction('registerRestoreOldStyle', false);
unexportedRuntimeFunction('hideEverythingExceptGivenElement', false);
unexportedRuntimeFunction('restoreHiddenElements', false);
unexportedRuntimeFunction('setLetterbox', false);
unexportedRuntimeFunction('currentFullscreenStrategy', false);
unexportedRuntimeFunction('restoreOldWindowedStyle', false);
unexportedRuntimeFunction('softFullscreenResizeWebGLRenderTarget', false);
unexportedRuntimeFunction('doRequestFullscreen', false);
unexportedRuntimeFunction('fillPointerlockChangeEventData', false);
unexportedRuntimeFunction('registerPointerlockChangeEventCallback', false);
unexportedRuntimeFunction('registerPointerlockErrorEventCallback', false);
unexportedRuntimeFunction('requestPointerLock', false);
unexportedRuntimeFunction('fillVisibilityChangeEventData', false);
unexportedRuntimeFunction('registerVisibilityChangeEventCallback', false);
unexportedRuntimeFunction('registerTouchEventCallback', false);
unexportedRuntimeFunction('fillGamepadEventData', false);
unexportedRuntimeFunction('registerGamepadEventCallback', false);
unexportedRuntimeFunction('registerBeforeUnloadEventCallback', false);
unexportedRuntimeFunction('fillBatteryEventData', false);
unexportedRuntimeFunction('battery', false);
unexportedRuntimeFunction('registerBatteryEventCallback', false);
unexportedRuntimeFunction('setCanvasElementSize', false);
unexportedRuntimeFunction('getCanvasElementSize', false);
unexportedRuntimeFunction('demangle', false);
unexportedRuntimeFunction('demangleAll', false);
unexportedRuntimeFunction('jsStackTrace', false);
unexportedRuntimeFunction('stackTrace', false);
unexportedRuntimeFunction('getEnvStrings', false);
unexportedRuntimeFunction('checkWasiClock', false);
unexportedRuntimeFunction('flush_NO_FILESYSTEM', false);
unexportedRuntimeFunction('dlopenMissingError', false);
unexportedRuntimeFunction('setImmediateWrapped', false);
unexportedRuntimeFunction('clearImmediateWrapped', false);
unexportedRuntimeFunction('polyfillSetImmediate', false);
unexportedRuntimeFunction('uncaughtExceptionCount', false);
unexportedRuntimeFunction('exceptionLast', false);
unexportedRuntimeFunction('exceptionCaught', false);
unexportedRuntimeFunction('ExceptionInfo', false);
unexportedRuntimeFunction('exception_addRef', false);
unexportedRuntimeFunction('exception_decRef', false);
unexportedRuntimeFunction('Browser', false);
unexportedRuntimeFunction('setMainLoop', false);
unexportedRuntimeFunction('wget', false);
unexportedRuntimeFunction('FS', false);
unexportedRuntimeFunction('MEMFS', false);
unexportedRuntimeFunction('TTY', false);
unexportedRuntimeFunction('PIPEFS', false);
unexportedRuntimeFunction('SOCKFS', false);
unexportedRuntimeFunction('_setNetworkCallback', false);
unexportedRuntimeFunction('tempFixedLengthArray', false);
unexportedRuntimeFunction('miniTempWebGLFloatBuffers', false);
unexportedRuntimeFunction('heapObjectForWebGLType', false);
unexportedRuntimeFunction('heapAccessShiftForWebGLHeap', false);
unexportedRuntimeFunction('GL', false);
unexportedRuntimeFunction('emscriptenWebGLGet', false);
unexportedRuntimeFunction('computeUnpackAlignedImageSize', false);
unexportedRuntimeFunction('emscriptenWebGLGetTexPixelData', false);
unexportedRuntimeFunction('emscriptenWebGLGetUniform', false);
unexportedRuntimeFunction('webglGetUniformLocation', false);
unexportedRuntimeFunction('webglPrepareUniformLocationsBeforeFirstUse', false);
unexportedRuntimeFunction('webglGetLeftBracePos', false);
unexportedRuntimeFunction('emscriptenWebGLGetVertexAttrib', false);
unexportedRuntimeFunction('writeGLArray', false);
unexportedRuntimeFunction('AL', false);
unexportedRuntimeFunction('SDL_unicode', false);
unexportedRuntimeFunction('SDL_ttfContext', false);
unexportedRuntimeFunction('SDL_audio', false);
unexportedRuntimeFunction('SDL', false);
unexportedRuntimeFunction('SDL_gfx', false);
unexportedRuntimeFunction('GLUT', false);
unexportedRuntimeFunction('EGL', false);
unexportedRuntimeFunction('GLFW_Window', false);
unexportedRuntimeFunction('GLFW', false);
unexportedRuntimeFunction('GLEW', false);
unexportedRuntimeFunction('IDBStore', false);
unexportedRuntimeFunction('runAndAbortIfError', false);
unexportedRuntimeSymbol('ALLOC_NORMAL', false);
unexportedRuntimeSymbol('ALLOC_STACK', false);

var calledRun;

/**
 * @constructor
 * @this {ExitStatus}
 */
function ExitStatus(status) {
  this.name = "ExitStatus";
  this.message = "Program terminated with exit(" + status + ")";
  this.status = status;
}

var calledMain = false;

dependenciesFulfilled = function runCaller() {
  // If run has never been called, and we should call run (INVOKE_RUN is true, and Module.noInitialRun is not false)
  if (!calledRun) run();
  if (!calledRun) dependenciesFulfilled = runCaller; // try this again later, after new deps are fulfilled
};

function stackCheckInit() {
  // This is normally called automatically during __wasm_call_ctors but need to
  // get these values before even running any of the ctors so we call it redundantly
  // here.
  _emscripten_stack_init();
  // TODO(sbc): Move writeStackCookie to native to to avoid this.
  writeStackCookie();
}

/** @type {function(Array=)} */
function run(args) {
  args = args || arguments_;

  if (runDependencies > 0) {
    return;
  }

    stackCheckInit();

  preRun();

  // a preRun added a dependency, run will be called later
  if (runDependencies > 0) {
    return;
  }

  function doRun() {
    // run may have just been called through dependencies being fulfilled just in this very frame,
    // or while the async setStatus time below was happening
    if (calledRun) return;
    calledRun = true;
    Module['calledRun'] = true;

    if (ABORT) return;

    initRuntime();

    if (Module['onRuntimeInitialized']) Module['onRuntimeInitialized']();

    assert(!Module['_main'], 'compiled without a main, but one is present. if you added it from JS, use Module["onRuntimeInitialized"]');

    postRun();
  }

  if (Module['setStatus']) {
    Module['setStatus']('Running...');
    setTimeout(function() {
      setTimeout(function() {
        Module['setStatus']('');
      }, 1);
      doRun();
    }, 1);
  } else
  {
    doRun();
  }
  checkStackCookie();
}
Module['run'] = run;

function checkUnflushedContent() {
  // Compiler settings do not allow exiting the runtime, so flushing
  // the streams is not possible. but in ASSERTIONS mode we check
  // if there was something to flush, and if so tell the user they
  // should request that the runtime be exitable.
  // Normally we would not even include flush() at all, but in ASSERTIONS
  // builds we do so just for this check, and here we see if there is any
  // content to flush, that is, we check if there would have been
  // something a non-ASSERTIONS build would have not seen.
  // How we flush the streams depends on whether we are in SYSCALLS_REQUIRE_FILESYSTEM=0
  // mode (which has its own special function for this; otherwise, all
  // the code is inside libc)
  var oldOut = out;
  var oldErr = err;
  var has = false;
  out = err = (x) => {
    has = true;
  }
  try { // it doesn't matter if it fails
    flush_NO_FILESYSTEM();
  } catch(e) {}
  out = oldOut;
  err = oldErr;
  if (has) {
    warnOnce('stdio streams had content in them that was not flushed. you should set EXIT_RUNTIME to 1 (see the FAQ), or make sure to emit a newline when you printf etc.');
    warnOnce('(this may also be due to not including full filesystem support - try building with -sFORCE_FILESYSTEM)');
  }
}

/** @param {boolean|number=} implicit */
function exit(status, implicit) {
  EXITSTATUS = status;

  checkUnflushedContent();

  // if exit() was called explicitly, warn the user if the runtime isn't actually being shut down
  if (keepRuntimeAlive() && !implicit) {
    var msg = 'program exited (with status: ' + status + '), but EXIT_RUNTIME is not set, so halting execution but not exiting the runtime or preventing further async execution (build with EXIT_RUNTIME=1, if you want a true shutdown)';
    err(msg);
  }

  procExit(status);
}

function procExit(code) {
  EXITSTATUS = code;
  if (!keepRuntimeAlive()) {
    if (Module['onExit']) Module['onExit'](code);
    ABORT = true;
  }
  quit_(code, new ExitStatus(code));
}

if (Module['preInit']) {
  if (typeof Module['preInit'] == 'function') Module['preInit'] = [Module['preInit']];
  while (Module['preInit'].length > 0) {
    Module['preInit'].pop()();
  }
}

run();





// Bindings utilities

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function WrapperObject() {
}
WrapperObject.prototype = Object.create(WrapperObject.prototype);
WrapperObject.prototype.constructor = WrapperObject;
WrapperObject.prototype.__class__ = WrapperObject;
WrapperObject.__cache__ = {};
Module['WrapperObject'] = WrapperObject;

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant)
    @param {*=} __class__ */
function getCache(__class__) {
  return (__class__ || WrapperObject).__cache__;
}
Module['getCache'] = getCache;

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant)
    @param {*=} __class__ */
function wrapPointer(ptr, __class__) {
  var cache = getCache(__class__);
  var ret = cache[ptr];
  if (ret) return ret;
  ret = Object.create((__class__ || WrapperObject).prototype);
  ret.ptr = ptr;
  return cache[ptr] = ret;
}
Module['wrapPointer'] = wrapPointer;

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function castObject(obj, __class__) {
  return wrapPointer(obj.ptr, __class__);
}
Module['castObject'] = castObject;

Module['NULL'] = wrapPointer(0);

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function destroy(obj) {
  if (!obj['__destroy__']) throw 'Error: Cannot destroy object. (Did you create it yourself?)';
  obj['__destroy__']();
  // Remove from cache, so the object can be GC'd and refs added onto it released
  delete getCache(obj.__class__)[obj.ptr];
}
Module['destroy'] = destroy;

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function compare(obj1, obj2) {
  return obj1.ptr === obj2.ptr;
}
Module['compare'] = compare;

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function getPointer(obj) {
  return obj.ptr;
}
Module['getPointer'] = getPointer;

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function getClass(obj) {
  return obj.__class__;
}
Module['getClass'] = getClass;

// Converts big (string or array) values into a C-style storage, in temporary space

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
var ensureCache = {
  buffer: 0,  // the main buffer of temporary storage
  // TODO: ensureCache seems to have issues increasing its size;
  // `Module['_free']` is `undefined`
  size: 1024,   // the size of buffer
  pos: 0,    // the next free offset in buffer
  temps: [], // extra allocations
  needed: 0, // the total size we need next time

  prepare: function() {
    if (ensureCache.needed) {
      // clear the temps
      for (var i = 0; i < ensureCache.temps.length; i++) {
        Module['_free'](ensureCache.temps[i]);
      }
      ensureCache.temps.length = 0;
      // prepare to allocate a bigger buffer
      Module['_free'](ensureCache.buffer);
      ensureCache.buffer = 0;
      ensureCache.size += ensureCache.needed;
      // clean up
      ensureCache.needed = 0;
    }
    if (!ensureCache.buffer) { // happens first time, or when we need to grow
      ensureCache.size += 128; // heuristic, avoid many small grow events
      ensureCache.buffer = Module['_malloc'](ensureCache.size);
      assert(ensureCache.buffer);
    }
    ensureCache.pos = 0;
  },
  alloc: function(array, view) {
    assert(ensureCache.buffer);
    var bytes = view.BYTES_PER_ELEMENT;
    var len = array.length * bytes;
    len = (len + 7) & -8; // keep things aligned to 8 byte boundaries
    var ret;
    if (ensureCache.pos + len >= ensureCache.size) {
      // we failed to allocate in the buffer, ensureCache time around :(
      assert(len > 0); // null terminator, at least
      ensureCache.needed += len;
      ret = Module['_malloc'](len);
      ensureCache.temps.push(ret);
    } else {
      // we can allocate in the buffer
      ret = ensureCache.buffer + ensureCache.pos;
      ensureCache.pos += len;
    }
    return ret;
  },
  copy: function(array, view, offset) {
    offset >>>= 0;
    var bytes = view.BYTES_PER_ELEMENT;
    switch (bytes) {
      case 2: offset >>>= 1; break;
      case 4: offset >>>= 2; break;
      case 8: offset >>>= 3; break;
    }
    for (var i = 0; i < array.length; i++) {
      view[offset + i] = array[i];
    }
  },
};

/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureString(value) {
  if (typeof value === 'string') {
    var intArray = intArrayFromString(value);
    var offset = ensureCache.alloc(intArray, HEAP8);
    ensureCache.copy(intArray, HEAP8, offset);
    return offset;
  }
  return value;
}
/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureUint8(value) {
  switch (typeof value) {
    case "string":
      value = intArrayFromString(value);
      break;
    case "object":
      break;
    default:
      throw new Error(`unsupported value type: ${value.constructor.name}`);
  }

  var offset = ensureCache.alloc(value, HEAPU8);
  ensureCache.copy(value, HEAPU8, offset);
  return offset;
}
/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureInt8(value) {
  if (typeof value === 'object') {
    var offset = ensureCache.alloc(value, HEAP8);
    ensureCache.copy(value, HEAP8, offset);
    return offset;
  }
  return value;
}
/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureInt16(value) {
  if (typeof value === 'object') {
    var offset = ensureCache.alloc(value, HEAP16);
    ensureCache.copy(value, HEAP16, offset);
    return offset;
  }
  return value;
}
/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureInt32(value) {
  if (typeof value === 'object') {
    var offset = ensureCache.alloc(value, HEAP32);
    ensureCache.copy(value, HEAP32, offset);
    return offset;
  }
  return value;
}
/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureFloat32(value) {
  if (typeof value === 'object') {
    var offset = ensureCache.alloc(value, HEAPF32);
    ensureCache.copy(value, HEAPF32, offset);
    return offset;
  }
  return value;
}
/** @suppress {duplicate} (TODO: avoid emitting this multiple times, it is redundant) */
function ensureFloat64(value) {
  if (typeof value === 'object') {
    var offset = ensureCache.alloc(value, HEAPF64);
    ensureCache.copy(value, HEAPF64, offset);
    return offset;
  }
  return value;
}


// VoidPtr
/** @suppress {undefinedVars, duplicate} @this{Object} */function VoidPtr() { throw "cannot construct a VoidPtr, no constructor in IDL" }
VoidPtr.prototype = Object.create(WrapperObject.prototype);
VoidPtr.prototype.constructor = VoidPtr;
VoidPtr.prototype.__class__ = VoidPtr;
VoidPtr.__cache__ = {};
Module['VoidPtr'] = VoidPtr;

  VoidPtr.prototype['__destroy__'] = VoidPtr.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_VoidPtr___destroy___0(self);
};
// P1_Affine
/** @suppress {undefinedVars, duplicate} @this{Object} */function P1_Affine(input, len) {
  ensureCache.prepare();
  if (input && typeof input === 'object') {
    if (input instanceof Uint8Array) {
      var _input = ensureUint8(input);
      this.ptr = _emscripten_bind_P1_Affine_P1_Affine_2(_input, input.length);
      return;
    } else {
      input= input.ptr;
    }
  }
  else input = ensureUint8(input);
  if (len && typeof len === 'object') len = len.ptr;
  if (input === undefined) { this.ptr = _emscripten_bind_P1_Affine_P1_Affine_0(); getCache(P1_Affine)[this.ptr] = this;return }
  if (len === undefined) { this.ptr = _emscripten_bind_P1_Affine_P1_Affine_1(input); getCache(P1_Affine)[this.ptr] = this;return }
  this.ptr = _emscripten_bind_P1_Affine_P1_Affine_2(input, len);
  getCache(P1_Affine)[this.ptr] = this;
};;
P1_Affine.prototype = Object.create(WrapperObject.prototype);
P1_Affine.prototype.constructor = P1_Affine;
P1_Affine.prototype.__class__ = P1_Affine;
P1_Affine.__cache__ = {};
Module['P1_Affine'] = P1_Affine;

P1_Affine.prototype['dup'] = P1_Affine.prototype.dup = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P1_Affine_dup_0(self), P1_Affine);
};;

P1_Affine.prototype['to_jacobian'] = P1_Affine.prototype.to_jacobian = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P1_Affine_to_jacobian_0(self), P1);
};;

P1_Affine.prototype['serialize'] = P1_Affine.prototype.serialize = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(96), HEAPU8);
  _emscripten_bind_P1_Affine_serialize_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 96));
  // TODO: free `out`?
};;

P1_Affine.prototype['compress'] = P1_Affine.prototype.compress = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(48), HEAPU8);
  _emscripten_bind_P1_Affine_compress_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 48));
  // TODO: free `out`?
};;

P1_Affine.prototype['on_curve'] = P1_Affine.prototype.on_curve = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P1_Affine_on_curve_0(self));
};;

P1_Affine.prototype['in_group'] = P1_Affine.prototype.in_group = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P1_Affine_in_group_0(self));
};;

P1_Affine.prototype['is_inf'] = P1_Affine.prototype.is_inf = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P1_Affine_is_inf_0(self));
};;

P1_Affine.prototype['is_equal'] = P1_Affine.prototype.is_equal = /** @suppress {undefinedVars, duplicate} @this{Object} */function(p) {
  var self = this.ptr;
  if (p && typeof p === 'object') p = p.ptr;
  return !!(_emscripten_bind_P1_Affine_is_equal_1(self, p));
};;

P1_Affine.prototype['core_verify'] = P1_Affine.prototype.core_verify = /** @suppress {undefinedVars, duplicate} @this{Object} */function(pk, hash_or_encode, msg, msg_len, DST, aug, aug_len) {
  var self = this.ptr;
  ensureCache.prepare();
  if (pk && typeof pk === 'object') pk = pk.ptr;
  if (hash_or_encode && typeof hash_or_encode === 'object') hash_or_encode = hash_or_encode.ptr;
  if (msg && typeof msg === 'object') msg = msg.ptr;
  else msg = ensureString(msg);
  if (msg_len && typeof msg_len === 'object') msg_len = msg_len.ptr;
  if (DST && typeof DST === 'object') DST = DST.ptr;
  else DST = ensureString(DST);
  if (aug && typeof aug === 'object') aug = aug.ptr;
  else aug = ensureString(aug);
  if (aug_len && typeof aug_len === 'object') aug_len = aug_len.ptr;
  return _emscripten_bind_P1_Affine_core_verify_7(self, pk, hash_or_encode, msg, msg_len, DST, aug, aug_len);
};;

  P1_Affine.prototype['__destroy__'] = P1_Affine.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_P1_Affine___destroy___0(self);
};
// P1
/** @suppress {undefinedVars, duplicate} @this{Object} */function P1(input, len) {
  ensureCache.prepare();
  if (input instanceof Uint8Array) {
    var _input = ensureUint8(input);
    this.ptr = _emscripten_bind_P1_P1_2(_input, input.length);
    getCache(P1)[this.ptr] = this;
    return;
  }
  if (input && typeof input === 'object') input = input.ptr;
  if (len && typeof len === 'object') len = len.ptr;
  if (input === undefined) { this.ptr = _emscripten_bind_P1_P1_0(); getCache(P1)[this.ptr] = this;return }
  if (len === undefined) { this.ptr = _emscripten_bind_P1_P1_1(input); getCache(P1)[this.ptr] = this;return }
  this.ptr = _emscripten_bind_P1_P1_2(input, len);
  getCache(P1)[this.ptr] = this;
};;
P1.prototype = Object.create(WrapperObject.prototype);
P1.prototype.constructor = P1;
P1.prototype.__class__ = P1;
P1.__cache__ = {};
Module['P1'] = P1;

P1.generator = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  return wrapPointer(_emscripten_bind_P1_P1_generator_0(), P1);
};;

P1.prototype['dup'] = P1.prototype.dup = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P1_dup_0(self), P1);
};;

P1.prototype['to_affine'] = P1.prototype.to_affine = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P1_to_affine_0(self), P1_Affine);
};;

P1.prototype['serialize'] = P1.prototype.serialize = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(96), HEAPU8);
  _emscripten_bind_P1_serialize_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 96));
  // TODO: free `out`?
};;

P1.prototype['compress'] = P1.prototype.compress = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(48), HEAPU8);
  _emscripten_bind_P1_compress_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 48));
  // TODO: free `out`?
};;

P1.prototype['on_curve'] = P1.prototype.on_curve = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P1_on_curve_0(self));
};;

P1.prototype['in_group'] = P1.prototype.in_group = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P1_in_group_0(self));
};;

P1.prototype['is_inf'] = P1.prototype.is_inf = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P1_is_inf_0(self));
};;

P1.prototype['is_equal'] = P1.prototype.is_equal = /** @suppress {undefinedVars, duplicate} @this{Object} */function(p) {
  var self = this.ptr;
  if (p && typeof p === 'object') p = p.ptr;
  return !!(_emscripten_bind_P1_is_equal_1(self, p));
};;

P1.prototype['aggregate'] = P1.prototype.aggregate = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
  var self = this.ptr;
  if (input && typeof input === 'object') input = input.ptr;
  _emscripten_bind_P1_aggregate_1(self, input);
};;

P1.prototype['sign_with'] = P1.prototype.sign_with = /** @suppress {undefinedVars, duplicate} @this{Object} */function(sk) {
  var self = this.ptr;
  if (sk && typeof sk === 'object') sk = sk.ptr;
  return wrapPointer(_emscripten_bind_P1_sign_with_1(self, sk), P1);
};;

P1.prototype['hash_to'] = P1.prototype.hash_to = /** @suppress {undefinedVars, duplicate} @this{Object} */function(msg, DST, aug) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_dst] = ensureRef(DST);
  const [_msg, msg_len] = ensureRef(msg);
  const [_aug, aug_len] = ensureRef(aug);

  return wrapPointer(_emscripten_bind_P1_hash_to_5(self, _msg, msg_len, _dst, _aug, aug_len), P1);
};;

P1.prototype['encode_to'] = P1.prototype.encode_to = /** @suppress {undefinedVars, duplicate} @this{Object} */function(msg, DST, aug) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_dst] = ensureRef(DST);
  const [_msg, msg_len] = ensureRef(msg);
  const [_aug, aug_len] = ensureRef(aug);

  return wrapPointer(_emscripten_bind_P1_encode_to_5(self, _msg, msg_len, _dst, _aug, aug_len), P1);
};;

P1.prototype['mult'] = P1.prototype.mult = /** @suppress {undefinedVars, duplicate} @this{Object} */function(scalar) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_scalar, scalar_len] = ensureRef(scalar);
  return wrapPointer(_emscripten_bind_P1_mult_1(self, _scalar, scalar_len * 8), P1);
};;

P1.prototype['neg'] = P1.prototype.neg = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P1_neg_0(self), P1);
};;

P1.prototype['cneg'] = P1.prototype.cneg = /** @suppress {undefinedVars, duplicate} @this{Object} */function(flag) {
  var self = this.ptr;
  if (flag && typeof flag === 'object') flag = flag.ptr;
  return wrapPointer(_emscripten_bind_P1_cneg_1(self, flag), P1);
};;

P1.prototype['add'] = P1.prototype.add = /** @suppress {undefinedVars, duplicate} @this{Object} */function(a) {
  var self = this.ptr;
  switch(a.constructor) {
    case P1:
      return wrapPointer(_emscripten_bind_P1_add_1__P1(self, a.ptr), P1);
    case P1_Affine:
      return wrapPointer(_emscripten_bind_P1_add_1__P1_Affine(self, a.ptr), P1);
    default:
      throw new Error("unsuported type for `a` arg");
  }
};;

P1.prototype['dbl'] = P1.prototype.dbl = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P1_dbl_0(self), P1);
};;

  P1.prototype['__destroy__'] = P1.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_P1___destroy___0(self);
};
// P2_Affine
/** @suppress {undefinedVars, duplicate} @this{Object} */function P2_Affine(input, len) {
  ensureCache.prepare();
  if (input && typeof input === 'object') {
    if (input instanceof Uint8Array) {
      var _input = ensureUint8(input);
      this.ptr = _emscripten_bind_P2_Affine_P2_Affine_2(_input, input.length);
      return;
    } else {
      input= input.ptr;
    }
  }
  else input = ensureUint8(input);
  if (input && typeof input === 'object') input = input.ptr;
  if (len && typeof len === 'object') len = len.ptr;
  if (input === undefined) { this.ptr = _emscripten_bind_P2_Affine_P2_Affine_0(); getCache(P2_Affine)[this.ptr] = this;return }
  if (len === undefined) { this.ptr = _emscripten_bind_P2_Affine_P2_Affine_1(input); getCache(P2_Affine)[this.ptr] = this;return }
  this.ptr = _emscripten_bind_P2_Affine_P2_Affine_2(input, len);
  getCache(P2_Affine)[this.ptr] = this;
};;
P2_Affine.prototype = Object.create(WrapperObject.prototype);
P2_Affine.prototype.constructor = P2_Affine;
P2_Affine.prototype.__class__ = P2_Affine;
P2_Affine.__cache__ = {};
Module['P2_Affine'] = P2_Affine;

P2_Affine.prototype['dup'] = P2_Affine.prototype.dup = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P2_Affine_dup_0(self), P2_Affine);
};;

P2_Affine.prototype['to_jacobian'] = P2_Affine.prototype.to_jacobian = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P2_Affine_to_jacobian_0(self), P2);
};;

P2_Affine.prototype['serialize'] = P2_Affine.prototype.serialize = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(192), HEAPU8);
  _emscripten_bind_P2_Affine_serialize_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 192));
  // TODO: free `out`?
};;

P2_Affine.prototype['compress'] = P2_Affine.prototype.compress = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(96), HEAPU8);
  _emscripten_bind_P2_Affine_compress_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 96));
};;

P2_Affine.prototype['on_curve'] = P2_Affine.prototype.on_curve = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P2_Affine_on_curve_0(self));
};;

P2_Affine.prototype['in_group'] = P2_Affine.prototype.in_group = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P2_Affine_in_group_0(self));
};;

P2_Affine.prototype['is_inf'] = P2_Affine.prototype.is_inf = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P2_Affine_is_inf_0(self));
};;

P2_Affine.prototype['is_equal'] = P2_Affine.prototype.is_equal = /** @suppress {undefinedVars, duplicate} @this{Object} */function(p) {
  var self = this.ptr;
  if (p && typeof p === 'object') p = p.ptr;
  return !!(_emscripten_bind_P2_Affine_is_equal_1(self, p));
};;

P2_Affine.prototype['core_verify'] = P2_Affine.prototype.core_verify = /** @suppress {undefinedVars, duplicate} @this{Object} */function(pk, hash_or_encode, msg, msg_len, DST, aug, aug_len) {
  var self = this.ptr;
  ensureCache.prepare();
  if (pk && typeof pk === 'object') pk = pk.ptr;
  if (hash_or_encode && typeof hash_or_encode === 'object') hash_or_encode = hash_or_encode.ptr;
  if (msg && typeof msg === 'object') msg = msg.ptr;
  else msg = ensureString(msg);
  if (msg_len && typeof msg_len === 'object') msg_len = msg_len.ptr;
  if (DST && typeof DST === 'object') DST = DST.ptr;
  else DST = ensureString(DST);
  if (aug && typeof aug === 'object') aug = aug.ptr;
  else aug = ensureString(aug);
  if (aug_len && typeof aug_len === 'object') aug_len = aug_len.ptr;
  return _emscripten_bind_P2_Affine_core_verify_7(self, pk, hash_or_encode, msg, msg_len, DST, aug, aug_len);
};;

  P2_Affine.prototype['__destroy__'] = P2_Affine.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_P2_Affine___destroy___0(self);
};
// P2
/** @suppress {undefinedVars, duplicate} @this{Object} */function P2(input, len) {
  ensureCache.prepare();
  if (input instanceof Uint8Array) {
    var _input = ensureUint8(input);
    this.ptr = _emscripten_bind_P2_P2_2(_input, input.length);
    getCache(P2)[this.ptr] = this;
    return;
  }
  if (input && typeof input === 'object') input = input.ptr;
  else input = ensureString(input);
  if (len && typeof len === 'object') len = len.ptr;
  if (input === undefined) { this.ptr = _emscripten_bind_P2_P2_0(); getCache(P2)[this.ptr] = this;return }
  if (len === undefined) { this.ptr = _emscripten_bind_P2_P2_1(input); getCache(P2)[this.ptr] = this;return }
  this.ptr = _emscripten_bind_P2_P2_2(input, len);
  getCache(P2)[this.ptr] = this;
};;
P2.prototype = Object.create(WrapperObject.prototype);
P2.prototype.constructor = P2;
P2.prototype.__class__ = P2;
P2.__cache__ = {};
Module['P2'] = P2;

P2.generator = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  return wrapPointer(_emscripten_bind_P2_P2_generator_0(), P2);
};;

P2.prototype['dup'] = P2.prototype.dup = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P2_dup_0(self), P2);
};;

P2.prototype['to_affine'] = P2.prototype.to_affine = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P2_to_affine_0(self), P2_Affine);
};;

P2.prototype['serialize'] = P2.prototype.serialize = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(192), HEAPU8);
  _emscripten_bind_P2_serialize_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 192));
};;

P2.prototype['compress'] = P2.prototype.compress = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(96), HEAPU8);
  _emscripten_bind_P2_compress_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 96));
};;

P2.prototype['on_curve'] = P2.prototype.on_curve = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P2_on_curve_0(self));
};;

P2.prototype['in_group'] = P2.prototype.in_group = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P2_in_group_0(self));
};;

P2.prototype['is_inf'] = P2.prototype.is_inf = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_P2_is_inf_0(self));
};;

P2.prototype['is_equal'] = P2.prototype.is_equal = /** @suppress {undefinedVars, duplicate} @this{Object} */function(p) {
  var self = this.ptr;
  if (p && typeof p === 'object') p = p.ptr;
  return !!(_emscripten_bind_P2_is_equal_1(self, p));
};;

P2.prototype['aggregate'] = P2.prototype.aggregate = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
  var self = this.ptr;
  if (input && typeof input === 'object') input = input.ptr;
  _emscripten_bind_P2_aggregate_1(self, input);
};;

P2.prototype['sign_with'] = P2.prototype.sign_with = /** @suppress {undefinedVars, duplicate} @this{Object} */function(scalar) {
  var self = this.ptr;
  if (scalar && typeof scalar === 'object') scalar = scalar.ptr;
  return wrapPointer(_emscripten_bind_P2_sign_with_1(self, scalar), P2);
};;

P2.prototype['hash_to'] = P2.prototype.hash_to = /** @suppress {undefinedVars, duplicate} @this{Object} */function(msg, DST, aug) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_dst] = ensureRef(DST);
  const [_msg, msg_len] = ensureRef(msg);
  const [_aug, aug_len] = ensureRef(aug);

  return wrapPointer(_emscripten_bind_P2_hash_to_5(self, _msg, msg_len, _dst, _aug, aug_len), P2);
};;

P2.prototype['encode_to'] = P2.prototype.encode_to = /** @suppress {undefinedVars, duplicate} @this{Object} */function(msg, DST, aug) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_dst] = ensureRef(DST);
  const [_msg, msg_len] = ensureRef(msg);
  const [_aug, aug_len] = ensureRef(aug);

  return wrapPointer(_emscripten_bind_P2_encode_to_5(self, _msg, msg_len, _dst, _aug, aug_len), P2);
};;

P2.prototype['mult'] = P2.prototype.mult = /** @suppress {undefinedVars, duplicate} @this{Object} */function(scalar) {
  var self = this.ptr;
  const [_scalar, scalar_len] = ensureRef(scalar);
  return wrapPointer(_emscripten_bind_P2_mult_1(self, _scalar, scalar_len * 8), P2);
};;

P2.prototype['neg'] = P2.prototype.neg = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P2_neg_0(self), P2);
};;

P2.prototype['cneg'] = P2.prototype.cneg = /** @suppress {undefinedVars, duplicate} @this{Object} */function(flag) {
  var self = this.ptr;
  if (flag && typeof flag === 'object') flag = flag.ptr;
  return wrapPointer(_emscripten_bind_P2_cneg_1(self, flag), P2);
};;

P2.prototype['add'] = P2.prototype.add = /** @suppress {undefinedVars, duplicate} @this{Object} */function(a) {
  var self = this.ptr;
  switch(a.constructor) {
    case P2:
      return wrapPointer(_emscripten_bind_P2_add_1__P2(self, a.ptr), P2);
    case P2_Affine:
      return wrapPointer(_emscripten_bind_P2_add_1__P2_Affine(self, a.ptr), P2);
    default:
      throw new Error("unsuported type for `a` arg");
  }
};;

P2.prototype['dbl'] = P2.prototype.dbl = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_P2_dbl_0(self), P2);
};;

  P2.prototype['__destroy__'] = P2.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_P2___destroy___0(self);
};
// SecretKey
/** @suppress {undefinedVars, duplicate} @this{Object} */function SecretKey() {
  this.ptr = _emscripten_bind_SecretKey_SecretKey_0();
  getCache(SecretKey)[this.ptr] = this;
};;
SecretKey.prototype = Object.create(WrapperObject.prototype);
SecretKey.prototype.constructor = SecretKey;
SecretKey.prototype.__class__ = SecretKey;
SecretKey.__cache__ = {};
Module['SecretKey'] = SecretKey;

SecretKey.prototype['keygen'] = SecretKey.prototype.keygen = /** @suppress {undefinedVars, duplicate} @this{Object} */function(ikm, info) {
  var self = this.ptr;
  var ikm_len = 0;
  ensureCache.prepare();
  if (ikm && typeof ikm === 'object') {
    if (ikm instanceof Uint8Array) {
      ikm_len = ikm.length;
      ikm = ensureUint8(ikm, ikm.length);
    } else {
      throw new Error("unsupported ikm type");
    }
  } else {
    ikm_len = ikm.length;
    ikm = ensureString(ikm);
  }
  if (info && typeof info === 'object') info = info.ptr;
  else info = ensureString(info || "");
  _emscripten_bind_SecretKey_keygen_3(self, ikm, ikm_len, info);
};;

SecretKey.prototype['from_bendian'] = SecretKey.prototype.from_bendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
  var self = this.ptr;
  ensureCache.prepare();
  if (typeof input == 'object') { input = ensureInt8(input); }
  _emscripten_bind_SecretKey_from_bendian_1(self, input);
};;

SecretKey.prototype['from_lendian'] = SecretKey.prototype.from_lendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
  var self = this.ptr;
  ensureCache.prepare();
  if (typeof input == 'object') { input = ensureInt8(input); }
  _emscripten_bind_SecretKey_from_lendian_1(self, input);
};;

SecretKey.prototype['to_bendian'] = SecretKey.prototype.to_bendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(32), HEAPU8);
  // var out = _malloc(32);
  _emscripten_bind_SecretKey_to_bendian_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 32));
  // TODO: free `out`?
};;

SecretKey.prototype['to_lendian'] = SecretKey.prototype.to_lendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  ensureCache.prepare();
  var out = ensureCache.alloc(new Uint8Array(32), HEAPU8);
  _emscripten_bind_SecretKey_to_lendian_1(self, out);
  return new Uint8Array(HEAPU8.subarray(out, out + 32));
  // TODO: free `out`?
};;

  SecretKey.prototype['__destroy__'] = SecretKey.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_SecretKey___destroy___0(self);
};

function unexpectedTypeError(expected, actual) {
  return new Error(`expected ${expected}, got: ${actual}`);
}

// PT
/** @suppress {undefinedVars, duplicate} @this{Object} */function PT(p, q) {
  if (q === undefined) {
    switch (p.constructor.name) {
      case "P1_Affine":
        this.ptr = _emscripten_bind_PT_PT__P1_Affine_1(p.ptr)
        break;
      case "P2_Affine":
        this.ptr = _emscripten_bind_PT_PT__P2_Affine_1(p.ptr)
        break;
      default:
        throw new Error(`unsupported type: ${p.constructor.name}`);
    }
    getCache(PT)[this.ptr] = this
    return
  }

  if (p.constructor === P1_Affine) {
    if (q.constructor !== P2_Affine) {
      throw unexpectedTypeError(P1_Affine, q.constructor);
    }
    this.ptr = _emscripten_bind_PT_PT_2(p.ptr, q.ptr);
  } else {
    if (q.constructor !== P1_Affine) {
      throw unexpectedTypeError(P2_Affine, q.constructor);
    }
    this.ptr = _emscripten_bind_PT_PT_2(q.ptr, p.ptr);
  }
  getCache(PT)[this.ptr] = this;
};;
PT.prototype = Object.create(WrapperObject.prototype);
PT.prototype.constructor = PT;
PT.prototype.__class__ = PT;
PT.__cache__ = {};
Module['PT'] = PT;

PT.prototype['dup'] = PT.prototype.dup = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_PT_dup_0(self), PT);
};;

PT.prototype['is_one'] = PT.prototype.is_one = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_PT_is_one_0(self));
};;

PT.prototype['is_equal'] = PT.prototype.is_equal = /** @suppress {undefinedVars, duplicate} @this{Object} */function(p) {
  var self = this.ptr;
  if (p && typeof p === 'object') p = p.ptr;
  return !!(_emscripten_bind_PT_is_equal_1(self, p));
};;

PT.prototype['sqr'] = PT.prototype.sqr = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_PT_sqr_0(self), PT);
};;

PT.prototype['mul'] = PT.prototype.mul = /** @suppress {undefinedVars, duplicate} @this{Object} */function(p) {
  var self = this.ptr;
  if (p && typeof p === 'object') p = p.ptr;
  return wrapPointer(_emscripten_bind_PT_mul_1(self, p), PT);
};;

PT.prototype['final_exp'] = PT.prototype.final_exp = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_PT_final_exp_0(self), PT);
};;

PT.prototype['in_group'] = PT.prototype.in_group = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return !!(_emscripten_bind_PT_in_group_0(self));
};;

PT.prototype['finalverify'] = PT.prototype.finalverify = /** @suppress {undefinedVars, duplicate} @this{Object} */function(gt1, gt2) {
  var self = this.ptr;
  if (gt1 && typeof gt1 === 'object') gt1 = gt1.ptr;
  if (gt2 && typeof gt2 === 'object') gt2 = gt2.ptr;
  return !!(_emscripten_bind_PT_finalverify_2(self, gt1, gt2));
};;

PT.prototype['one'] = PT.prototype.one = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  return wrapPointer(_emscripten_bind_PT_one_0(self), PT);
};;

  PT.prototype['__destroy__'] = PT.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_PT___destroy___0(self);
};
// Pairing
/** @suppress {undefinedVars, duplicate} @this{Object} */function Pairing(hash_or_encode, DST) {
  ensureCache.prepare();
  const [_dst, DST_len] = ensureRef(DST);
  this.ptr = _emscripten_bind_Pairing_Pairing_3(hash_or_encode, _dst, DST_len);
  getCache(Pairing)[this.ptr] = this;
};;
Pairing.prototype = Object.create(WrapperObject.prototype);
Pairing.prototype.constructor = Pairing;
Pairing.prototype.__class__ = Pairing;
Pairing.__cache__ = {};
Module['Pairing'] = Pairing;

Pairing.prototype['aggregate'] = Pairing.prototype.aggregate = /** @suppress {undefinedVars, duplicate} @this{Object} */function(pk, sig, msg, aug) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_msg, msg_len] = ensureRef(msg);
  const [_aug, aug_len] = ensureRef(aug);

  if (pk.constructor === P1_Affine) {
    pk = pk.ptr;

    if (sig.constructor !== P2_Affine) {
      throw new Error("`pk` is P1_Affine, expected `sig` to be P2_Affine; received: " + sig.constructor.name);
    }
    sig = sig.ptr;
    return _emscripten_bind_Pairing_aggregate_6__P1_P2(self, pk, sig, _msg, msg_len, _aug, aug_len);
  } else {
    pk = pk.ptr;

    if (sig.constructor !== P1_Affine) {
      throw new Error("`pk` is P2_Affine, expected `sig` to be P1_Affine; received: " + sig.constructor.name);
    }
    sig = sig.ptr;
    return _emscripten_bind_Pairing_aggregate_6__P2_P1(self, pk, sig, _msg, msg_len, _aug, aug_len);
  }
};;

Pairing.prototype['mul_n_aggregate'] = Pairing.prototype.mul_n_aggregate = /** @suppress {undefinedVars, duplicate} @this{Object} */function(pk, sig, scalar, msg, aug) {
  var self = this.ptr;
  ensureCache.prepare();

  const [_scalar, scalar_len] = ensureRef(scalar);
  const [_msg, msg_len] = ensureRef(msg);
  const [_aug, aug_len] = ensureRef(aug);

  if (pk.constructor === P1_Affine) {
    pk = pk.ptr;

    if (sig.constructor !== P2_Affine) {
      throw new Error("`pk` is P1_Affine, expected `sig` to be P2_Affine; received: " + sig.constructor.name);
    }
    sig = sig.ptr;
    return _emscripten_bind_Pairing_mul_n_aggregate_8__P1_P2(self, pk, sig, _scalar, scalar_len, _msg, msg_len, _aug, aug_len);
  } else {
    pk = pk.ptr;

    if (sig.constructor !== P1_Affine) {
      throw new Error("`pk` is P2_Affine, expected `sig` to be P1_Affine; received: " + sig.constructor.name);
    }
    sig = sig.ptr;
    return _emscripten_bind_Pairing_mul_n_aggregate_8__P2_P1(self, pk, sig, _scalar, scalar_len, _msg, msg_len, _aug, aug_len);
  }
};;

Pairing.prototype['commit'] = Pairing.prototype.commit = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_Pairing_commit_0(self);
};;

Pairing.prototype['merge'] = Pairing.prototype.merge = /** @suppress {undefinedVars, duplicate} @this{Object} */function(ctx) {
  var self = this.ptr;
  if (ctx && typeof ctx === 'object') ctx = ctx.ptr;
  return _emscripten_bind_Pairing_merge_1(self, ctx);
};;

Pairing.prototype['finalverify'] = Pairing.prototype.finalverify = /** @suppress {undefinedVars, duplicate} @this{Object} */function(sig) {
  var self = this.ptr;
  if (sig && typeof sig === 'object') sig = sig.ptr;
  return !!(_emscripten_bind_Pairing_finalverify_1(self, sig));
};;

Pairing.prototype['__destroy__'] = Pairing.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
  var self = this.ptr;
  _emscripten_bind_Pairing___destroy___0(self);
};

// G1
/** @suppress {undefinedVars, duplicate} @this{Object} */function G1() {
  return wrapPointer(_emscripten_bind_G1_0(), P1);
};;
Module['G1'] = G1;

// G2
/** @suppress {undefinedVars, duplicate} @this{Object} */function G2() {
  return wrapPointer(_emscripten_bind_G2_0(), P2);
};;
Module['G2'] = G2;

function ensureRef(val) {
  // TODO: something else (nullptr)?
  if (typeof val === "undefined") {
    return [ensureString(""), 0];
  }

  switch (val.constructor.name) {
    case "String":
      return [ensureString(val), val.length];
    case "Buffer":
      // NB: intentional fallthrough.
    case "Uint8Array":
      return [ensureUint8(val), val.length];
    // TODO:
    // case BigInt:
    default:
      throw new Error(`unsupported type for \`val\`: ${val.constructor.name}`);
  }
}

const initialized = new Promise(resolve => {
  Module['onRuntimeInitialized'] = function() {
    resolve();
  };
});

function onInitialized(callback) {
  initialized.then(callback).catch((e) => {
    throw e;
  });
}

module.exports = {
  initialized,
  onInitialized,
  // BLS12_381_G1,
  // BLS12_381_NEG_G1,
  // BLS12_381_G2,
  // BLS12_381_NEG_G2,
  SecretKey,
  P1,
  P2,
  P1_Affine,
  P2_Affine,
  PT,
  Pairing,
  G1,
  G2,
};
